// @audius/eth/dist/cjs/trusted-notifier-manager/trustedNotifierManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TrustedNotifierManager = void 0;
const abi_1 = require("./abi");
const constants_1 = require("./constants");
class TrustedNotifierManager {
    constructor(client, { address } = {}) {
        this.getLatestNotifierID = () => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getLatestNotifierID'
        });
        this.getNotifierForID = ({ id }) => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getNotifierForID',
            args: [id]
        });
        this.getNotifierForWallet = ({ address }) => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getNotifierForWallet',
            args: [address]
        });
        this.getNotifierForEndpoint = ({ endpoint }) => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getNotifierForEndpoint',
            args: [endpoint]
        });
        this.getNotifierForEmail = ({ email }) => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getNotifierForEmail',
            args: [email]
        });
        this.client = client;
        this.address = address !== null && address !== void 0 ? address : constants_1.TRUSTED_NOTIFIER_MANAGER_CONTRACT_ADDRESS;
    }
}
exports.TrustedNotifierManager = TrustedNotifierManager;
