// @audius/eth/dist/esm/staking/staking.js
import { abi } from './abi';
import { STAKING_CONTRACT_ADDRESS } from './constants';
export class Staking {
    constructor(client, { address } = {}) {
        this.token = () => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'token'
        });
        this.totalStaked = () => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'totalStaked'
        });
        this.supportsHistory = () => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'supportsHistory'
        });
        this.totalStakedFor = ({ account }) => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'totalStakedFor',
            args: [account]
        });
        this.totalStakedAt = ({ blockNumber }) => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'totalStakedAt',
            args: [blockNumber]
        });
        this.totalStakedForAt = ({ account, blockNumber }) => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'totalStakedForAt',
            args: [account, blockNumber]
        });
        this.isStaker = ({ account }) => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'isStaker',
            args: [account]
        });
        this.lastClaimedFor = ({ account }) => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'lastClaimedFor',
            args: [account]
        });
        this.getDelegateManagerAddress = () => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'getDelegateManagerAddress'
        });
        this.getClaimsManagerAddress = () => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'getClaimsManagerAddress'
        });
        this.getServiceProviderFactoryAddress = () => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'getServiceProviderFactoryAddress'
        });
        this.getGovernanceAddress = () => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'getGovernanceAddress'
        });
        this.client = client;
        this.address = address ?? STAKING_CONTRACT_ADDRESS;
    }
}
