// @injectivelabs/ts-types/dist/cjs/trade.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderSide = exports.OrderState = exports.TradeDirection = exports.TradeExecutionSide = exports.TradeExecutionType = void 0;
exports.TradeExecutionType = {
    Market: 'market',
    LimitFill: 'limitFill',
    LimitMatchRestingOrder: 'limitMatchRestingOrder',
    LimitMatchNewOrder: 'limitMatchNewOrder',
};
exports.TradeExecutionSide = {
    Maker: 'maker',
    Taker: 'taker',
};
exports.TradeDirection = {
    Buy: 'buy',
    Sell: 'sell',
    Long: 'long',
    Short: 'short',
};
exports.OrderState = {
    Unfilled: 'unfilled',
    Booked: 'booked',
    PartialFilled: 'partial_filled',
    PartiallyFilled: 'partially_filled',
    Filled: 'filled',
    Canceled: 'canceled',
    Triggered: 'triggered',
};
exports.OrderSide = {
    Unspecified: 'unspecified',
    Buy: 'buy',
    Sell: 'sell',
    StopBuy: 'stop_buy',
    StopSell: 'stop_sell',
    TakeBuy: 'take_buy',
    TakeSell: 'take_sell',
    BuyPO: 'buy_po',
    SellPO: 'sell_po',
    BuyAtomic: 'buy_atomic',
    SellAtomic: 'sell_atomic',
    Unrecognized: 'unrecognized',
};
