// @injectivelabs/sdk-ts/dist/cjs/client/chain/transformers/ChainGrpcCommonTransformer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainGrpcCommonTransformer = void 0;
const core_proto_ts_1 = require("@injectivelabs/core-proto-ts");
class ChainGrpcCommonTransformer {
    static grpcCoinToCoin(coin) {
        return {
            denom: coin.denom,
            amount: coin.amount,
        };
    }
    static pageRequestToGrpcPageRequest(pageRequest) {
        const paginationForRequest = core_proto_ts_1.CosmosBaseQueryV1Beta1Pagination.PageRequest.create();
        if (!pageRequest) {
            return;
        }
        if (pageRequest.key) {
            paginationForRequest.key = Buffer.from(pageRequest.key, 'base64');
        }
        if (pageRequest.limit !== undefined) {
            paginationForRequest.limit = pageRequest.limit.toString();
        }
        if (pageRequest.offset !== undefined) {
            paginationForRequest.offset = pageRequest.offset.toString();
        }
        if (pageRequest.reverse !== undefined) {
            paginationForRequest.reverse = pageRequest.reverse;
        }
        if (pageRequest.countTotal !== undefined) {
            paginationForRequest.countTotal = pageRequest.countTotal;
        }
        return paginationForRequest;
    }
    static paginationUint8ArrayToString(key) {
        if (!key) {
            return '';
        }
        if (key.constructor !== Uint8Array) {
            return key;
        }
        return Buffer.from(key).toString('base64');
    }
    static grpcPaginationToPagination(pagination) {
        return {
            total: pagination
                ? parseInt(ChainGrpcCommonTransformer.paginationUint8ArrayToString(pagination.total), 10)
                : 0,
            next: pagination
                ? ChainGrpcCommonTransformer.paginationUint8ArrayToString(pagination.nextKey)
                : '',
        };
    }
}
exports.ChainGrpcCommonTransformer = ChainGrpcCommonTransformer;
