// @injectivelabs/sdk-ts/dist/cjs/client/chain/transformers/ChainGrpcTxFeesTransformer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainGrpcTxFeesTransformer = void 0;
const numbers_js_1 = require("./../../../utils/numbers.js");
/**
 * @category Chain Grpc Transformer
 */
class ChainGrpcTxFeesTransformer {
    static moduleParamsResponseToModuleParams(response) {
        const params = response.params;
        return {
            maxGasWantedPerTx: params.maxGasWantedPerTx,
            highGasTxThreshold: params.highGasTxThreshold,
            minGasPriceForHighGasTx: params.minGasPriceForHighGasTx,
            mempool1559Enabled: params.mempool1559Enabled,
            minGasPrice: (0, numbers_js_1.denomAmountFromGrpcChainDenomAmount)(params.minGasPrice).toFixed(),
            defaultBaseFeeMultiplier: (0, numbers_js_1.denomAmountFromGrpcChainDenomAmount)(params.defaultBaseFeeMultiplier).toFixed(),
            maxBaseFeeMultiplier: (0, numbers_js_1.denomAmountFromGrpcChainDenomAmount)(params.maxBaseFeeMultiplier).toFixed(),
            resetInterval: params.resetInterval,
            maxBlockChangeRate: (0, numbers_js_1.denomAmountFromGrpcChainDenomAmount)(params.maxBlockChangeRate).toFixed(),
            targetBlockSpacePercentRate: (0, numbers_js_1.denomAmountFromGrpcChainDenomAmount)(params.targetBlockSpacePercentRate).toFixed(),
            recheckFeeLowBaseFee: (0, numbers_js_1.denomAmountFromGrpcChainDenomAmount)(params.recheckFeeLowBaseFee).toFixed(),
            recheckFeeHighBaseFee: (0, numbers_js_1.denomAmountFromGrpcChainDenomAmount)(params.recheckFeeHighBaseFee).toFixed(),
            recheckFeeBaseFeeThresholdMultiplier: (0, numbers_js_1.denomAmountFromGrpcChainDenomAmount)(params.recheckFeeBaseFeeThresholdMultiplier).toFixed(),
        };
    }
    static eipBaseFeeResponseToEipBaseFee(response) {
        return {
            baseFee: response.baseFee
                ? (0, numbers_js_1.denomAmountFromGrpcChainDenomAmount)(response.baseFee.baseFee).toFixed()
                : undefined,
        };
    }
}
exports.ChainGrpcTxFeesTransformer = ChainGrpcTxFeesTransformer;
