// @injectivelabs/sdk-ts/dist/cjs/client/chain/transformers/ChainGrpcErc20Transformer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainGrpcErc20Transformer = void 0;
const ChainGrpcCommonTransformer_js_1 = require("./ChainGrpcCommonTransformer.js");
/**
 * @category Chain Grpc Transformer
 */
class ChainGrpcErc20Transformer {
    static grpcTokenPairToTokenPair(grpcTokenPair) {
        return {
            bankDenom: grpcTokenPair.bankDenom,
            erc20Address: grpcTokenPair.erc20Address,
        };
    }
    static paramsResponseToParams(response) {
        const params = response.params;
        return {
            denomCreationFee: params?.denomCreationFee,
        };
    }
    static tokenPairsResponseToTokenPairs(response) {
        return {
            tokenPairs: response.tokenPairs.map((tokenPair) => ({
                bankDenom: tokenPair.bankDenom,
                erc20Address: tokenPair.erc20Address,
            })),
            pagination: ChainGrpcCommonTransformer_js_1.ChainGrpcCommonTransformer.grpcPaginationToPagination(response.pagination),
        };
    }
}
exports.ChainGrpcErc20Transformer = ChainGrpcErc20Transformer;
