// @injectivelabs/core-proto-ts/esm/injective/downtimedetector/v1beta1/genesis.js
/* eslint-disable */
import _m0 from "protobufjs/minimal.js";
import { Timestamp } from "../../../google/protobuf/timestamp.js";
import { downtimeFromJSON, downtimeToJSON } from "./downtime_duration.js";
export const protobufPackage = "injective.downtimedetector.v1beta1";
function createBaseGenesisDowntimeEntry() {
    return { duration: 0, lastDowntime: undefined };
}
export const GenesisDowntimeEntry = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.duration !== 0) {
            writer.uint32(8).int32(message.duration);
        }
        if (message.lastDowntime !== undefined) {
            Timestamp.encode(toTimestamp(message.lastDowntime), writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGenesisDowntimeEntry();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.duration = reader.int32();
                    break;
                case 2:
                    message.lastDowntime = fromTimestamp(Timestamp.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            duration: isSet(object.duration) ? downtimeFromJSON(object.duration) : 0,
            lastDowntime: isSet(object.lastDowntime) ? fromJsonTimestamp(object.lastDowntime) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.duration !== undefined && (obj.duration = downtimeToJSON(message.duration));
        message.lastDowntime !== undefined && (obj.lastDowntime = message.lastDowntime.toISOString());
        return obj;
    },
    create(base) {
        return GenesisDowntimeEntry.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseGenesisDowntimeEntry();
        message.duration = object.duration ?? 0;
        message.lastDowntime = object.lastDowntime ?? undefined;
        return message;
    },
};
function createBaseGenesisState() {
    return { downtimes: [], lastBlockTime: undefined };
}
export const GenesisState = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.downtimes) {
            GenesisDowntimeEntry.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (message.lastBlockTime !== undefined) {
            Timestamp.encode(toTimestamp(message.lastBlockTime), writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGenesisState();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.downtimes.push(GenesisDowntimeEntry.decode(reader, reader.uint32()));
                    break;
                case 2:
                    message.lastBlockTime = fromTimestamp(Timestamp.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            downtimes: Array.isArray(object?.downtimes)
                ? object.downtimes.map((e) => GenesisDowntimeEntry.fromJSON(e))
                : [],
            lastBlockTime: isSet(object.lastBlockTime) ? fromJsonTimestamp(object.lastBlockTime) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.downtimes) {
            obj.downtimes = message.downtimes.map((e) => e ? GenesisDowntimeEntry.toJSON(e) : undefined);
        }
        else {
            obj.downtimes = [];
        }
        message.lastBlockTime !== undefined && (obj.lastBlockTime = message.lastBlockTime.toISOString());
        return obj;
    },
    create(base) {
        return GenesisState.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseGenesisState();
        message.downtimes = object.downtimes?.map((e) => GenesisDowntimeEntry.fromPartial(e)) || [];
        message.lastBlockTime = object.lastBlockTime ?? undefined;
        return message;
    },
};
function toTimestamp(date) {
    const seconds = Math.trunc(date.getTime() / 1_000).toString();
    const nanos = (date.getTime() % 1_000) * 1_000_000;
    return { seconds, nanos };
}
function fromTimestamp(t) {
    let millis = Number(t.seconds) * 1_000;
    millis += t.nanos / 1_000_000;
    return new Date(millis);
}
function fromJsonTimestamp(o) {
    if (o instanceof Date) {
        return o;
    }
    else if (typeof o === "string") {
        return new Date(o);
    }
    else {
        return fromTimestamp(Timestamp.fromJSON(o));
    }
}
function isSet(value) {
    return value !== null && value !== undefined;
}
