// @injectivelabs/core-proto-ts/esm/injective/downtimedetector/v1beta1/query.js
/* eslint-disable */
import { grpc } from "@injectivelabs/grpc-web";
import { BrowserHeaders } from "browser-headers";
import _m0 from "protobufjs/minimal.js";
import { Duration } from "../../../google/protobuf/duration.js";
import { downtimeFromJSON, downtimeToJSON } from "./downtime_duration.js";
export const protobufPackage = "injective.downtimedetector.v1beta1";
function createBaseRecoveredSinceDowntimeOfLengthRequest() {
    return { downtime: 0, recovery: undefined };
}
export const RecoveredSinceDowntimeOfLengthRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.downtime !== 0) {
            writer.uint32(8).int32(message.downtime);
        }
        if (message.recovery !== undefined) {
            Duration.encode(message.recovery, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRecoveredSinceDowntimeOfLengthRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.downtime = reader.int32();
                    break;
                case 2:
                    message.recovery = Duration.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            downtime: isSet(object.downtime) ? downtimeFromJSON(object.downtime) : 0,
            recovery: isSet(object.recovery) ? Duration.fromJSON(object.recovery) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.downtime !== undefined && (obj.downtime = downtimeToJSON(message.downtime));
        message.recovery !== undefined && (obj.recovery = message.recovery ? Duration.toJSON(message.recovery) : undefined);
        return obj;
    },
    create(base) {
        return RecoveredSinceDowntimeOfLengthRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseRecoveredSinceDowntimeOfLengthRequest();
        message.downtime = object.downtime ?? 0;
        message.recovery = (object.recovery !== undefined && object.recovery !== null)
            ? Duration.fromPartial(object.recovery)
            : undefined;
        return message;
    },
};
function createBaseRecoveredSinceDowntimeOfLengthResponse() {
    return { successfullyRecovered: false };
}
export const RecoveredSinceDowntimeOfLengthResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.successfullyRecovered === true) {
            writer.uint32(8).bool(message.successfullyRecovered);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRecoveredSinceDowntimeOfLengthResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.successfullyRecovered = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            successfullyRecovered: isSet(object.successfullyRecovered) ? Boolean(object.successfullyRecovered) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        message.successfullyRecovered !== undefined && (obj.successfullyRecovered = message.successfullyRecovered);
        return obj;
    },
    create(base) {
        return RecoveredSinceDowntimeOfLengthResponse.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseRecoveredSinceDowntimeOfLengthResponse();
        message.successfullyRecovered = object.successfullyRecovered ?? false;
        return message;
    },
};
export class QueryClientImpl {
    rpc;
    constructor(rpc) {
        this.rpc = rpc;
        this.RecoveredSinceDowntimeOfLength = this.RecoveredSinceDowntimeOfLength.bind(this);
    }
    RecoveredSinceDowntimeOfLength(request, metadata) {
        return this.rpc.unary(QueryRecoveredSinceDowntimeOfLengthDesc, RecoveredSinceDowntimeOfLengthRequest.fromPartial(request), metadata);
    }
}
export const QueryDesc = { serviceName: "injective.downtimedetector.v1beta1.Query" };
export const QueryRecoveredSinceDowntimeOfLengthDesc = {
    methodName: "RecoveredSinceDowntimeOfLength",
    service: QueryDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return RecoveredSinceDowntimeOfLengthRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = RecoveredSinceDowntimeOfLengthResponse.decode(data);
            return {
                ...value,
                toObject() {
                    return value;
                },
            };
        },
    },
};
export class GrpcWebImpl {
    host;
    options;
    constructor(host, options) {
        this.host = host;
        this.options = options;
    }
    unary(methodDesc, _request, metadata) {
        const request = { ..._request, ...methodDesc.requestType };
        const maybeCombinedMetadata = metadata && this.options.metadata
            ? new BrowserHeaders({ ...this.options?.metadata.headersMap, ...metadata?.headersMap })
            : metadata || this.options.metadata;
        return new Promise((resolve, reject) => {
            grpc.unary(methodDesc, {
                request,
                host: this.host,
                metadata: maybeCombinedMetadata,
                transport: this.options.transport,
                debug: this.options.debug,
                onEnd: function (response) {
                    if (response.status === grpc.Code.OK) {
                        resolve(response.message.toObject());
                    }
                    else {
                        const err = new GrpcWebError(response.statusMessage, response.status, response.trailers);
                        reject(err);
                    }
                },
            });
        });
    }
}
var tsProtoGlobalThis = (() => {
    if (typeof globalThis !== "undefined") {
        return globalThis;
    }
    if (typeof self !== "undefined") {
        return self;
    }
    if (typeof window !== "undefined") {
        return window;
    }
    if (typeof global !== "undefined") {
        return global;
    }
    throw "Unable to locate global object";
})();
function isSet(value) {
    return value !== null && value !== undefined;
}
export class GrpcWebError extends tsProtoGlobalThis.Error {
    code;
    metadata;
    constructor(message, code, metadata) {
        super(message);
        this.code = code;
        this.metadata = metadata;
    }
}
