// @audius/eth/dist/cjs/claims-manager/claimsManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClaimsManager = void 0;
const abi_1 = require("./abi");
const constants_1 = require("./constants");
class ClaimsManager {
    constructor(client, { address } = {}) {
        // Get the duration of a funding round in blocks
        this.getFundingRoundBlockDiff = () => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getFundingRoundBlockDiff'
        });
        // Get the last block where a funding round was initiated
        this.getLastFundedBlock = () => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getLastFundedBlock'
        });
        // Get the amount funded per round in wei
        this.getFundsPerRound = () => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getFundsPerRound'
        });
        // total amount claimed in the current round
        this.getTotalClaimedInRound = () => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getTotalClaimedInRound'
        });
        // Get the Governance address
        this.getGovernanceAddress = () => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getGovernanceAddress'
        });
        // Get the ServiceProviderFactory address
        this.getServiceProviderFactoryAddress = () => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getServiceProviderFactoryAddress'
        });
        // Get the DelegateManager address
        this.getDelegateManagerAddress = () => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getDelegateManagerAddress'
        });
        // Get the Staking address
        this.getStakingAddress = () => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getStakingAddress'
        });
        this.claimPending = ({ address }) => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'claimPending',
            args: [address]
        });
        this.client = client;
        this.address = address !== null && address !== void 0 ? address : constants_1.CLAIMS_MANAGER_CONTRACT_ADDRESS;
    }
}
exports.ClaimsManager = ClaimsManager;
