// @injectivelabs/utils/dist/cjs/helpers.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStdFee = exports.getStdFeeFromString = exports.getStdFeeFromObject = exports.getStdFeeForToken = exports.splitArrayToChunks = exports.splitArrayToChunksThrow = exports.awaitForAll = exports.awaitAll = exports.sleep = void 0;
const constants_js_1 = require("./constants.js");
const BigNumber_js_1 = __importStar(require("./classes/BigNumber.js"));
const sleep = (timeout) => new Promise((resolve) => setTimeout(resolve, timeout));
exports.sleep = sleep;
/**
 * When we want to execute the promises in batch
 */
const awaitAll = async (array, callback) => await Promise.all(array.map(async (item) => await callback(item)));
exports.awaitAll = awaitAll;
/**
 * When we want to execute the promises one by one
 * and not all in batch as with await Promise.all()
 */
const awaitForAll = async (array, callback) => {
    const result = [];
    for (let i = 0; i < array.length; i += 1) {
        try {
            result.push(await callback(array[i]));
        }
        catch {
            //
        }
    }
    return result;
};
exports.awaitForAll = awaitForAll;
const splitArrayToChunksThrow = ({ array, filter, chunkSize, }) => {
    const chunks = [];
    const chunkSizeInNumber = Number(chunkSize);
    if (isNaN(chunkSizeInNumber)) {
        throw new Error('Invalid chunk size, must be a valid number');
    }
    for (let i = 0; i < array.length; i += chunkSizeInNumber) {
        const chunk = array.slice(i, i + chunkSizeInNumber);
        if (filter) {
            chunks.push(chunk.filter(filter));
        }
        else {
            chunks.push(chunk);
        }
    }
    return chunks;
};
exports.splitArrayToChunksThrow = splitArrayToChunksThrow;
const splitArrayToChunks = ({ array, filter, chunkSize, }) => {
    try {
        return (0, exports.splitArrayToChunksThrow)({ array, chunkSize, filter });
    }
    catch {
        return [array];
    }
};
exports.splitArrayToChunks = splitArrayToChunks;
const getStdFeeForToken = (token = { denom: 'inj', decimals: 18 }, gasPrice, gasLimit) => {
    const gasPriceInBase = gasPrice || (0, BigNumber_js_1.toHumanReadable)(constants_js_1.DEFAULT_GAS_PRICE, 18);
    const gasPriceScaled = (0, BigNumber_js_1.toChainFormat)(gasPriceInBase, token.decimals).toFixed(0);
    const gasNormalized = new BigNumber_js_1.default(gasLimit || constants_js_1.DEFAULT_GAS_LIMIT).toFixed(0);
    return {
        amount: [
            {
                denom: token.denom,
                amount: new BigNumber_js_1.default(gasPriceScaled).times(gasNormalized).toFixed(),
            },
        ],
        gas: (gasLimit || constants_js_1.DEFAULT_GAS_LIMIT).toString(),
    };
};
exports.getStdFeeForToken = getStdFeeForToken;
const getStdFeeFromObject = (args) => {
    if (!args) {
        return (0, constants_js_1.getDefaultStdFee)();
    }
    const { payer, granter, feePayer, gasPrice = constants_js_1.DEFAULT_GAS_PRICE, gas = constants_js_1.DEFAULT_GAS_LIMIT.toString(), } = args;
    const gasNormalized = (0, BigNumber_js_1.toBigNumber)(gas).toFixed(0);
    const gasPriceNormalized = (0, BigNumber_js_1.toBigNumber)(gasPrice).toFixed(0);
    return {
        amount: [
            {
                denom: 'inj',
                amount: (0, BigNumber_js_1.toBigNumber)(gasNormalized).times(gasPriceNormalized).toFixed(),
            },
        ],
        gas: (0, BigNumber_js_1.toBigNumber)(gasNormalized).toFixed(),
        payer /** for Web3Gateway fee delegation */,
        granter,
        feePayer,
    };
};
exports.getStdFeeFromObject = getStdFeeFromObject;
const getStdFeeFromString = (gasPrice) => {
    const matchResult = gasPrice.match(/^([0-9.]+)([a-zA-Z][a-zA-Z0-9/:._-]*)$/);
    if (!matchResult) {
        throw new Error('Invalid gas price string');
    }
    const [_, amount] = matchResult;
    const gas = (0, BigNumber_js_1.toChainFormat)(amount, 18).dividedBy(constants_js_1.DEFAULT_GAS_PRICE).toFixed(0);
    return (0, exports.getStdFeeFromObject)({ gas, gasPrice: constants_js_1.DEFAULT_GAS_PRICE });
};
exports.getStdFeeFromString = getStdFeeFromString;
const getStdFee = (args) => {
    if (!args) {
        return (0, constants_js_1.getDefaultStdFee)();
    }
    if (typeof args === 'string') {
        return (0, exports.getStdFeeFromString)(args);
    }
    return (0, exports.getStdFeeFromObject)({ ...args });
};
exports.getStdFee = getStdFee;
