// @injectivelabs/utils/dist/cjs/time.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEndDateStringFromTimeInSeconds = exports.pastDays = exports.past24Hours = exports.todayInSeconds = exports.tomorrow = exports.getUTCDateFromTimestamp = exports.convertTimestampToMilliseconds = exports.getSecondsInDay = void 0;
const BigNumber_js_1 = __importDefault(require("./classes/BigNumber.js"));
const getSecondsInDay = () => new BigNumber_js_1.default(60 * 60 * 24);
exports.getSecondsInDay = getSecondsInDay;
const convertTimestampToMilliseconds = (timestamp) => {
    const timestampInBigNumber = new BigNumber_js_1.default(timestamp);
    if (timestamp.toString().length > 13) {
        return timestampInBigNumber
            .precision(13, BigNumber_js_1.default.ROUND_HALF_UP)
            .toNumber();
    }
    if (timestamp.toString().length < 13) {
        const trailingZeros = 13 - timestamp.toString().length;
        return timestampInBigNumber.times(10 ** trailingZeros).toNumber();
    }
    return timestampInBigNumber.toNumber();
};
exports.convertTimestampToMilliseconds = convertTimestampToMilliseconds;
const getUTCDateFromTimestamp = (timestamp) => {
    const date = new Date((0, exports.convertTimestampToMilliseconds)(timestamp));
    return `${date.getUTCDate()}-${date.getUTCMonth() + 1}-${date.getUTCFullYear()}`;
};
exports.getUTCDateFromTimestamp = getUTCDateFromTimestamp;
const tomorrow = () => new BigNumber_js_1.default(Math.floor(new Date().valueOf() / 1000) + 3600 * 24);
exports.tomorrow = tomorrow;
const todayInSeconds = () => Math.floor(Date.now() / 1000);
exports.todayInSeconds = todayInSeconds;
const past24Hours = () => new BigNumber_js_1.default((0, exports.todayInSeconds)()).minus((0, exports.getSecondsInDay)()).toNumber();
exports.past24Hours = past24Hours;
const pastDays = (day = 1) => new BigNumber_js_1.default((0, exports.todayInSeconds)()).minus((0, exports.getSecondsInDay)().times(day)).toNumber();
exports.pastDays = pastDays;
const getEndDateStringFromTimeInSeconds = (timeInSeconds) => {
    const currentDate = new Date(timeInSeconds.toNumber() * 1000);
    return currentDate.toLocaleString('en-us');
};
exports.getEndDateStringFromTimeInSeconds = getEndDateStringFromTimeInSeconds;
