// @injectivelabs/core-proto-ts/esm/injective/types/v1beta1/indexer.js
/* eslint-disable */
import Long from "long";
import _m0 from "protobufjs/minimal.js";
export const protobufPackage = "injective.types.v1beta1";
function createBaseTxResult() {
    return { height: "0", txIndex: 0, msgIndex: 0, ethTxIndex: 0, failed: false, gasUsed: "0", cumulativeGasUsed: "0" };
}
export const TxResult = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.height !== "0") {
            writer.uint32(8).int64(message.height);
        }
        if (message.txIndex !== 0) {
            writer.uint32(16).uint32(message.txIndex);
        }
        if (message.msgIndex !== 0) {
            writer.uint32(24).uint32(message.msgIndex);
        }
        if (message.ethTxIndex !== 0) {
            writer.uint32(32).int32(message.ethTxIndex);
        }
        if (message.failed === true) {
            writer.uint32(40).bool(message.failed);
        }
        if (message.gasUsed !== "0") {
            writer.uint32(48).uint64(message.gasUsed);
        }
        if (message.cumulativeGasUsed !== "0") {
            writer.uint32(56).uint64(message.cumulativeGasUsed);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTxResult();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.height = longToString(reader.int64());
                    break;
                case 2:
                    message.txIndex = reader.uint32();
                    break;
                case 3:
                    message.msgIndex = reader.uint32();
                    break;
                case 4:
                    message.ethTxIndex = reader.int32();
                    break;
                case 5:
                    message.failed = reader.bool();
                    break;
                case 6:
                    message.gasUsed = longToString(reader.uint64());
                    break;
                case 7:
                    message.cumulativeGasUsed = longToString(reader.uint64());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            height: isSet(object.height) ? String(object.height) : "0",
            txIndex: isSet(object.txIndex) ? Number(object.txIndex) : 0,
            msgIndex: isSet(object.msgIndex) ? Number(object.msgIndex) : 0,
            ethTxIndex: isSet(object.ethTxIndex) ? Number(object.ethTxIndex) : 0,
            failed: isSet(object.failed) ? Boolean(object.failed) : false,
            gasUsed: isSet(object.gasUsed) ? String(object.gasUsed) : "0",
            cumulativeGasUsed: isSet(object.cumulativeGasUsed) ? String(object.cumulativeGasUsed) : "0",
        };
    },
    toJSON(message) {
        const obj = {};
        message.height !== undefined && (obj.height = message.height);
        message.txIndex !== undefined && (obj.txIndex = Math.round(message.txIndex));
        message.msgIndex !== undefined && (obj.msgIndex = Math.round(message.msgIndex));
        message.ethTxIndex !== undefined && (obj.ethTxIndex = Math.round(message.ethTxIndex));
        message.failed !== undefined && (obj.failed = message.failed);
        message.gasUsed !== undefined && (obj.gasUsed = message.gasUsed);
        message.cumulativeGasUsed !== undefined && (obj.cumulativeGasUsed = message.cumulativeGasUsed);
        return obj;
    },
    create(base) {
        return TxResult.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseTxResult();
        message.height = object.height ?? "0";
        message.txIndex = object.txIndex ?? 0;
        message.msgIndex = object.msgIndex ?? 0;
        message.ethTxIndex = object.ethTxIndex ?? 0;
        message.failed = object.failed ?? false;
        message.gasUsed = object.gasUsed ?? "0";
        message.cumulativeGasUsed = object.cumulativeGasUsed ?? "0";
        return message;
    },
};
function longToString(long) {
    return long.toString();
}
if (_m0.util.Long !== Long) {
    _m0.util.Long = Long;
    _m0.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
