// @injectivelabs/sdk-ts/dist/cjs/core/modules/exchange/msgs/MsgCancelPostOnlyModeV2.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const snakecase_keys_1 = __importDefault(require("snakecase-keys"));
const core_proto_ts_1 = require("@injectivelabs/core-proto-ts");
const MsgBase_js_1 = require("../../MsgBase.js");
const createMessage = (params) => {
    const message = core_proto_ts_1.InjectiveExchangeV2Tx.MsgCancelPostOnlyMode.create();
    message.sender = params.sender;
    return core_proto_ts_1.InjectiveExchangeV2Tx.MsgCancelPostOnlyMode.fromPartial(message);
};
/**
 * @category Messages
 */
class MsgCancelPostOnlyModeV2 extends MsgBase_js_1.MsgBase {
    static fromJSON(params) {
        return new MsgCancelPostOnlyModeV2(params);
    }
    toProto() {
        const { params: initialParams } = this;
        const params = {
            ...initialParams,
        };
        return createMessage(params);
    }
    toData() {
        const proto = this.toProto();
        return {
            '@type': '/injective.exchange.v2.MsgCancelPostOnlyMode',
            ...proto,
        };
    }
    toAmino() {
        const { params } = this;
        const msg = createMessage(params);
        const message = {
            ...(0, snakecase_keys_1.default)(msg),
        };
        return {
            type: 'exchange/MsgCancelPostOnlyMode',
            value: message,
        };
    }
    toWeb3Gw() {
        const amino = this.toAmino();
        const { value } = amino;
        return {
            '@type': '/injective.exchange.v2.MsgCancelPostOnlyMode',
            ...value,
        };
    }
    toEip712() {
        const amino = this.toAmino();
        const { type, value } = amino;
        const messageAdjusted = {
            ...value,
        };
        return {
            type,
            value: messageAdjusted,
        };
    }
    toEip712V2() {
        const web3gw = this.toWeb3Gw();
        const messageAdjusted = {
            ...web3gw,
        };
        return messageAdjusted;
    }
    toDirectSign() {
        const proto = this.toProto();
        return {
            type: '/injective.exchange.v2.MsgCancelPostOnlyMode',
            message: proto,
        };
    }
    toBinary() {
        return core_proto_ts_1.InjectiveExchangeV2Tx.MsgCancelPostOnlyMode.encode(this.toProto()).finish();
    }
}
exports.default = MsgCancelPostOnlyModeV2;
