// @injectivelabs/sdk-ts/dist/cjs/core/modules/exchange/msgs/MsgCreateBinaryOptionsLimitOrder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const snakecase_keys_1 = __importDefault(require("snakecase-keys"));
const utils_1 = require("@injectivelabs/utils");
const core_proto_ts_1 = require("@injectivelabs/core-proto-ts");
const MsgBase_js_1 = require("../../MsgBase.js");
const numbers_js_1 = require("../../../../utils/numbers.js");
const createLimitOrder = (params) => {
    const orderInfo = core_proto_ts_1.InjectiveExchangeV1Beta1Exchange.OrderInfo.create();
    orderInfo.subaccountId = params.subaccountId;
    orderInfo.feeRecipient = params.feeRecipient;
    orderInfo.price = params.price;
    orderInfo.quantity = params.quantity;
    if (params.cid) {
        orderInfo.cid = params.cid;
    }
    const derivativeOrder = core_proto_ts_1.InjectiveExchangeV1Beta1Exchange.DerivativeOrder.create();
    derivativeOrder.marketId = params.marketId;
    derivativeOrder.orderInfo = orderInfo;
    derivativeOrder.orderType = params.orderType;
    derivativeOrder.margin = params.margin;
    derivativeOrder.triggerPrice = params.triggerPrice || '0';
    const message = core_proto_ts_1.InjectiveExchangeV1Beta1Tx.MsgCreateBinaryOptionsLimitOrder.create();
    message.sender = params.injectiveAddress;
    message.order = derivativeOrder;
    return core_proto_ts_1.InjectiveExchangeV1Beta1Tx.MsgCreateBinaryOptionsLimitOrder.fromPartial(message);
};
/**
 * @category Messages
 */
class MsgCreateBinaryOptionsLimitOrder extends MsgBase_js_1.MsgBase {
    static fromJSON(params) {
        return new MsgCreateBinaryOptionsLimitOrder(params);
    }
    toProto() {
        const { params: initialParams } = this;
        const params = {
            ...initialParams,
            price: (0, utils_1.toChainFormat)(initialParams.price).toFixed(),
            margin: (0, utils_1.toChainFormat)(initialParams.margin).toFixed(),
            triggerPrice: (0, utils_1.toChainFormat)(initialParams.triggerPrice || 0).toFixed(),
            quantity: (0, utils_1.toChainFormat)(initialParams.quantity).toFixed(),
        };
        return createLimitOrder(params);
    }
    toData() {
        const proto = this.toProto();
        return {
            '@type': '/injective.exchange.v1beta1.MsgCreateBinaryOptionsLimitOrder',
            ...proto,
        };
    }
    toAmino() {
        const { params } = this;
        const order = createLimitOrder(params);
        const message = {
            ...(0, snakecase_keys_1.default)(order),
        };
        return {
            type: 'exchange/MsgCreateBinaryOptionsLimitOrder',
            value: message,
        };
    }
    toWeb3Gw() {
        const amino = this.toAmino();
        const { value } = amino;
        return {
            '@type': '/injective.exchange.v1beta1.MsgCreateBinaryOptionsLimitOrder',
            ...value,
        };
    }
    toEip712V2() {
        const { params } = this;
        const web3gw = this.toWeb3Gw();
        const order = web3gw.order;
        const messageAdjusted = {
            ...web3gw,
            order: {
                ...order,
                order_info: {
                    ...order.order_info,
                    price: (0, numbers_js_1.numberToCosmosSdkDecString)(params.price),
                    quantity: (0, numbers_js_1.numberToCosmosSdkDecString)(params.quantity),
                },
                margin: (0, numbers_js_1.numberToCosmosSdkDecString)(params.margin),
                trigger_price: (0, numbers_js_1.numberToCosmosSdkDecString)(params.triggerPrice || '0'),
                order_type: core_proto_ts_1.InjectiveExchangeV1Beta1Exchange.orderTypeToJSON(params.orderType),
            },
        };
        return messageAdjusted;
    }
    toEip712() {
        const { params } = this;
        const amino = this.toAmino();
        const { value, type } = amino;
        const messageAdjusted = {
            ...value,
            order: {
                ...value.order,
                order_info: {
                    ...value.order?.order_info,
                    price: (0, utils_1.toChainFormat)(params.price).toFixed(),
                    quantity: (0, utils_1.toChainFormat)(params.quantity).toFixed(),
                },
                margin: (0, utils_1.toChainFormat)(params.margin).toFixed(),
                trigger_price: (0, utils_1.toChainFormat)(params.triggerPrice || '0').toFixed(),
            },
        };
        return {
            type,
            value: messageAdjusted,
        };
    }
    toDirectSign() {
        const proto = this.toProto();
        return {
            type: '/injective.exchange.v1beta1.MsgCreateBinaryOptionsLimitOrder',
            message: proto,
        };
    }
    toBinary() {
        return core_proto_ts_1.InjectiveExchangeV1Beta1Tx.MsgCreateBinaryOptionsLimitOrder.encode(this.toProto()).finish();
    }
}
exports.default = MsgCreateBinaryOptionsLimitOrder;
