// @injectivelabs/sdk-ts/dist/cjs/core/modules/exchange/msgs/MsgSetDelegationTransferReceivers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const snakecase_keys_1 = __importDefault(require("snakecase-keys"));
const core_proto_ts_1 = require("@injectivelabs/core-proto-ts");
const MsgBase_js_1 = require("../../MsgBase.js");
/**
 * @category Messages
 */
class MsgSetDelegationTransferReceivers extends MsgBase_js_1.MsgBase {
    static fromJSON(params) {
        return new MsgSetDelegationTransferReceivers(params);
    }
    toProto() {
        const { params } = this;
        const message = core_proto_ts_1.InjectiveExchangeV2Tx.MsgSetDelegationTransferReceivers.create();
        message.sender = params.sender;
        message.receivers = params.receivers;
        return core_proto_ts_1.InjectiveExchangeV2Tx.MsgSetDelegationTransferReceivers.fromPartial(message);
    }
    toData() {
        const proto = this.toProto();
        return {
            '@type': '/injective.exchange.v2.MsgSetDelegationTransferReceivers',
            ...proto,
        };
    }
    toAmino() {
        const proto = this.toProto();
        const message = {
            ...(0, snakecase_keys_1.default)(proto),
        };
        return {
            type: 'exchange/MsgSetDelegationTransferReceivers',
            value: message,
        };
    }
    toWeb3Gw() {
        const amino = this.toAmino();
        const { value } = amino;
        return {
            '@type': '/injective.exchange.v2.MsgSetDelegationTransferReceivers',
            ...value,
        };
    }
    toDirectSign() {
        const proto = this.toProto();
        return {
            type: '/injective.exchange.v2.MsgSetDelegationTransferReceivers',
            message: proto,
        };
    }
    toBinary() {
        return core_proto_ts_1.InjectiveExchangeV2Tx.MsgSetDelegationTransferReceivers.encode(this.toProto()).finish();
    }
}
exports.default = MsgSetDelegationTransferReceivers;
