// @injectivelabs/sdk-ts/dist/cjs/core/modules/exchange/msgs/MsgUpdateDerivativeMarketV2.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const snakecase_keys_1 = __importDefault(require("snakecase-keys"));
const utils_1 = require("@injectivelabs/utils");
const core_proto_ts_1 = require("@injectivelabs/core-proto-ts");
const MsgBase_js_1 = require("../../MsgBase.js");
const numbers_js_1 = require("../../../../utils/numbers.js");
const createMessage = (params) => {
    const message = core_proto_ts_1.InjectiveExchangeV2Tx.MsgUpdateDerivativeMarket.create();
    message.admin = params.admin;
    message.marketId = params.marketId;
    message.newTicker = params.newTicker || '';
    message.newMinNotional = params.newMinNotional || '0';
    message.newMinPriceTickSize = params.newMinPriceTickSize || '0';
    message.newMinQuantityTickSize = params.newMinQuantityTickSize || '0';
    message.newInitialMarginRatio = params.newInitialMarginRatio || '0';
    message.newMaintenanceMarginRatio = params.newMaintenanceMarginRatio || '0';
    message.newReduceMarginRatio = params.newReduceMarginRatio || '0';
    return core_proto_ts_1.InjectiveExchangeV2Tx.MsgUpdateDerivativeMarket.fromPartial(message);
};
/**
 * @category Messages
 */
class MsgUpdateDerivativeMarketV2 extends MsgBase_js_1.MsgBase {
    static fromJSON(params) {
        return new MsgUpdateDerivativeMarketV2(params);
    }
    toProto() {
        const { params: initialParams } = this;
        const params = {
            ...initialParams,
            newMinNotional: (0, utils_1.toChainFormat)(initialParams.newMinNotional || '0').toFixed(),
            newMinQuantityTickSize: (0, utils_1.toChainFormat)(initialParams.newMinQuantityTickSize || '0').toFixed(),
            newMinPriceTickSize: (0, utils_1.toChainFormat)(initialParams.newMinPriceTickSize || '0').toFixed(),
            newInitialMarginRatio: (0, utils_1.toChainFormat)(initialParams.newInitialMarginRatio || '0').toFixed(),
            newMaintenanceMarginRatio: (0, utils_1.toChainFormat)(initialParams.newMaintenanceMarginRatio || '0').toFixed(),
            newReduceMarginRatio: (0, utils_1.toChainFormat)(initialParams.newReduceMarginRatio || '0').toFixed(),
        };
        return createMessage(params);
    }
    toData() {
        const proto = this.toProto();
        return {
            '@type': '/injective.exchange.v2.MsgUpdateDerivativeMarket',
            ...proto,
        };
    }
    toAmino() {
        const { params } = this;
        const msg = createMessage(params);
        const message = {
            ...(0, snakecase_keys_1.default)(msg),
        };
        return {
            type: 'exchange/MsgUpdateDerivativeMarket',
            value: message,
        };
    }
    toWeb3Gw() {
        const amino = this.toAmino();
        const { value } = amino;
        return {
            '@type': '/injective.exchange.v2.MsgUpdateDerivativeMarket',
            ...value,
        };
    }
    toEip712() {
        const amino = this.toAmino();
        const { type, value } = amino;
        const messageAdjusted = {
            ...value,
        };
        return {
            type,
            value: messageAdjusted,
        };
    }
    toEip712V2() {
        const { params } = this;
        const web3gw = this.toWeb3Gw();
        const messageAdjusted = {
            ...web3gw,
            new_min_price_tick_size: (0, numbers_js_1.numberToCosmosSdkDecString)(params.newMinPriceTickSize || '0'),
            new_min_quantity_tick_size: (0, numbers_js_1.numberToCosmosSdkDecString)(params.newMinQuantityTickSize || '0'),
            new_min_notional: (0, numbers_js_1.numberToCosmosSdkDecString)(params.newMinNotional || '0'),
            new_initial_margin_ratio: (0, numbers_js_1.numberToCosmosSdkDecString)(params.newInitialMarginRatio || '0'),
            new_maintenance_margin_ratio: (0, numbers_js_1.numberToCosmosSdkDecString)(params.newMaintenanceMarginRatio || '0'),
            new_reduce_margin_ratio: (0, numbers_js_1.numberToCosmosSdkDecString)(params.newReduceMarginRatio || '0'),
        };
        return messageAdjusted;
    }
    toDirectSign() {
        const proto = this.toProto();
        return {
            type: '/injective.exchange.v2.MsgUpdateDerivativeMarket',
            message: proto,
        };
    }
    toBinary() {
        return core_proto_ts_1.InjectiveExchangeV2Tx.MsgUpdateDerivativeMarket.encode(this.toProto()).finish();
    }
}
exports.default = MsgUpdateDerivativeMarketV2;
