// @audius/spl/dist/cjs/payment-router/PaymentRouterProgram.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentRouterProgram = void 0;
const anchor_1 = require("@coral-xyz/anchor");
const spl_token_1 = require("@solana/spl-token");
const web3_js_1 = require("@solana/web3.js");
const paymentRouter_1 = require("./paymentRouter");
class PaymentRouterProgram {
    static createPaymentRouterBalancePdaInstruction(paymentRouterPda, payer, programId = PaymentRouterProgram.programId) {
        return __awaiter(this, void 0, void 0, function* () {
            const program = new anchor_1.Program(paymentRouter_1.IDL, programId);
            return yield program.methods
                .createPaymentRouterBalancePda()
                .accounts({
                paymentRouterPda,
                payer,
                systemProgram: web3_js_1.SystemProgram.programId
            })
                .instruction();
        });
    }
    static createRouteInstruction({ sender, senderOwner, paymentRouterPdaBump, recipients, amounts, totalAmount, tokenProgramId = spl_token_1.TOKEN_PROGRAM_ID, programId = PaymentRouterProgram.programId }) {
        return __awaiter(this, void 0, void 0, function* () {
            const program = new anchor_1.Program(paymentRouter_1.IDL, programId, {});
            return yield program.methods
                .route(paymentRouterPdaBump, amounts.map((b) => new anchor_1.BN(b.toString())), new anchor_1.BN(totalAmount.toString()))
                .accounts({
                sender,
                senderOwner,
                splToken: tokenProgramId
            })
                .remainingAccounts(recipients.map((recipient) => ({
                pubkey: recipient,
                isSigner: false,
                isWritable: true
            })))
                .instruction();
        });
    }
}
PaymentRouterProgram.programId = new web3_js_1.PublicKey('paytYpX3LPN98TAeen6bFFeraGSuWnomZmCXjAsoqPa');
exports.PaymentRouterProgram = PaymentRouterProgram;
