// @audius/eth/dist/cjs/service-provider-factory/serviceProviderFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceProviderFactory = void 0;
const abi_1 = require("./abi");
const constants_1 = require("./constants");
class ServiceProviderFactory {
    constructor(client, { address } = {}) {
        this.getTotalServiceTypeProviders = ({ serviceType }) => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getTotalServiceTypeProviders',
            args: [serviceType]
        });
        this.getServiceEndpointInfo = ({ serviceType, index }) => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getServiceEndpointInfo',
            args: [serviceType, index]
        });
        this.client = client;
        this.address = address !== null && address !== void 0 ? address : constants_1.SERVICE_PROVIDER_FACTORY_CONTRACT_ADDRESS;
    }
}
exports.ServiceProviderFactory = ServiceProviderFactory;
