// @injectivelabs/sdk-ts/dist/cjs/utils/pagination.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchAllWithPagination = exports.grpcPagingToPaging = exports.grpcPaginationToPagination = exports.paginationUint8ArrayToString = exports.paginationRequestFromPagination = void 0;
const ChainGrpcCommonTransformer_js_1 = require("../client/chain/transformers/ChainGrpcCommonTransformer.js");
/**
 * @deprecated Use ChainGrpcCommonTransformer.pageRequestToGrpcPageRequest instead
 */
const paginationRequestFromPagination = (pagination) => {
    return ChainGrpcCommonTransformer_js_1.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequest(pagination);
};
exports.paginationRequestFromPagination = paginationRequestFromPagination;
/**
 * @deprecated Use ChainGrpcCommonTransformer.paginationUint8ArrayToString instead
 */
const paginationUint8ArrayToString = (key) => {
    return ChainGrpcCommonTransformer_js_1.ChainGrpcCommonTransformer.paginationUint8ArrayToString(key);
};
exports.paginationUint8ArrayToString = paginationUint8ArrayToString;
/**
 * @deprecated Use ChainGrpcCommonTransformer.grpcPaginationToPagination instead
 */
const grpcPaginationToPagination = (pagination) => {
    return ChainGrpcCommonTransformer_js_1.ChainGrpcCommonTransformer.grpcPaginationToPagination(pagination);
};
exports.grpcPaginationToPagination = grpcPaginationToPagination;
const grpcPagingToPaging = (pagination) => {
    if (!pagination) {
        return {
            to: 0,
            from: 0,
            total: 0,
        };
    }
    return {
        ...pagination,
        to: parseInt(pagination.to.toString() || '0', 10),
        from: parseInt(pagination.from.toString() || '0', 10),
        total: parseInt(pagination.total || '0', 10),
    };
};
exports.grpcPagingToPaging = grpcPagingToPaging;
const fetchAllWithPagination = async (args, method, result = []) => {
    let response = await method(args);
    if (!args) {
        return response;
    }
    const keys = Object.keys(response);
    const valueKey = keys.find((key) => key !== 'pagination');
    result.push(...response[valueKey]);
    const paginationOption = args;
    if (response.pagination.next) {
        return (0, exports.fetchAllWithPagination)({ ...paginationOption, key: response.pagination.next }, method, result);
    }
    return { [valueKey]: result, pagination: response.pagination };
};
exports.fetchAllWithPagination = fetchAllWithPagination;
