// @injectivelabs/core-proto-ts/cjs/injective/types/v1beta1/indexer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxResult = exports.protobufPackage = void 0;
/* eslint-disable */
var long_1 = __importDefault(require("long"));
var minimal_js_1 = __importDefault(require("protobufjs/minimal.js"));
exports.protobufPackage = "injective.types.v1beta1";
function createBaseTxResult() {
    return { height: "0", txIndex: 0, msgIndex: 0, ethTxIndex: 0, failed: false, gasUsed: "0", cumulativeGasUsed: "0" };
}
exports.TxResult = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.height !== "0") {
            writer.uint32(8).int64(message.height);
        }
        if (message.txIndex !== 0) {
            writer.uint32(16).uint32(message.txIndex);
        }
        if (message.msgIndex !== 0) {
            writer.uint32(24).uint32(message.msgIndex);
        }
        if (message.ethTxIndex !== 0) {
            writer.uint32(32).int32(message.ethTxIndex);
        }
        if (message.failed === true) {
            writer.uint32(40).bool(message.failed);
        }
        if (message.gasUsed !== "0") {
            writer.uint32(48).uint64(message.gasUsed);
        }
        if (message.cumulativeGasUsed !== "0") {
            writer.uint32(56).uint64(message.cumulativeGasUsed);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseTxResult();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.height = longToString(reader.int64());
                    break;
                case 2:
                    message.txIndex = reader.uint32();
                    break;
                case 3:
                    message.msgIndex = reader.uint32();
                    break;
                case 4:
                    message.ethTxIndex = reader.int32();
                    break;
                case 5:
                    message.failed = reader.bool();
                    break;
                case 6:
                    message.gasUsed = longToString(reader.uint64());
                    break;
                case 7:
                    message.cumulativeGasUsed = longToString(reader.uint64());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            height: isSet(object.height) ? String(object.height) : "0",
            txIndex: isSet(object.txIndex) ? Number(object.txIndex) : 0,
            msgIndex: isSet(object.msgIndex) ? Number(object.msgIndex) : 0,
            ethTxIndex: isSet(object.ethTxIndex) ? Number(object.ethTxIndex) : 0,
            failed: isSet(object.failed) ? Boolean(object.failed) : false,
            gasUsed: isSet(object.gasUsed) ? String(object.gasUsed) : "0",
            cumulativeGasUsed: isSet(object.cumulativeGasUsed) ? String(object.cumulativeGasUsed) : "0",
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.height !== undefined && (obj.height = message.height);
        message.txIndex !== undefined && (obj.txIndex = Math.round(message.txIndex));
        message.msgIndex !== undefined && (obj.msgIndex = Math.round(message.msgIndex));
        message.ethTxIndex !== undefined && (obj.ethTxIndex = Math.round(message.ethTxIndex));
        message.failed !== undefined && (obj.failed = message.failed);
        message.gasUsed !== undefined && (obj.gasUsed = message.gasUsed);
        message.cumulativeGasUsed !== undefined && (obj.cumulativeGasUsed = message.cumulativeGasUsed);
        return obj;
    },
    create: function (base) {
        return exports.TxResult.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b, _c, _d, _e, _f, _g;
        var message = createBaseTxResult();
        message.height = (_a = object.height) !== null && _a !== void 0 ? _a : "0";
        message.txIndex = (_b = object.txIndex) !== null && _b !== void 0 ? _b : 0;
        message.msgIndex = (_c = object.msgIndex) !== null && _c !== void 0 ? _c : 0;
        message.ethTxIndex = (_d = object.ethTxIndex) !== null && _d !== void 0 ? _d : 0;
        message.failed = (_e = object.failed) !== null && _e !== void 0 ? _e : false;
        message.gasUsed = (_f = object.gasUsed) !== null && _f !== void 0 ? _f : "0";
        message.cumulativeGasUsed = (_g = object.cumulativeGasUsed) !== null && _g !== void 0 ? _g : "0";
        return message;
    },
};
function longToString(long) {
    return long.toString();
}
if (minimal_js_1.default.util.Long !== long_1.default) {
    minimal_js_1.default.util.Long = long_1.default;
    minimal_js_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
