// @audius/eth/dist/cjs/service-type-manager/serviceTypeManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceTypeManager = void 0;
const abi_1 = require("./abi");
const constants_1 = require("./constants");
class ServiceTypeManager {
    constructor(client, { address } = {}) {
        this.getGovernanceAddress = () => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getGovernanceAddress'
        });
        this.getServiceTypeInfo = ({ serviceType }) => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getServiceTypeInfo',
            args: [serviceType]
        });
        this.getValidServiceTypes = () => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getValidServiceTypes'
        });
        this.serviceTypeIsValid = ({ serviceType }) => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'serviceTypeIsValid',
            args: [serviceType]
        });
        this.getVersion = ({ serviceType, versionIndex }) => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getVersion',
            args: [serviceType, versionIndex]
        });
        this.getCurrentVersion = ({ serviceType }) => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getCurrentVersion',
            args: [serviceType]
        });
        this.getNumberOfVersions = ({ serviceType }) => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getNumberOfVersions',
            args: [serviceType]
        });
        // TODO - writes
        this.setGovernanceAddress = null;
        this.addServiceType = null;
        this.removeServiceType = null;
        this.setServiceVersion = null;
        this.serviceVersionIsValid = null;
        this.client = client;
        this.address = address !== null && address !== void 0 ? address : constants_1.SERVICE_TYPE_MANAGER_CONTRACT_ADDRESS;
    }
}
exports.ServiceTypeManager = ServiceTypeManager;
