// @audius/eth/dist/esm/delegate-manager/delegateManager.js
import { abi } from './abi';
import { DELEGATE_MANAGER_CONTRACT_ADDRESS } from './constants';
export class DelegateManager {
    constructor(client, { address } = {}) {
        this.getIncreaseDelegatedStakeEvents = ({ fromBlock = BigInt(0), delegator, serviceProvider, increaseAmount }) => {
            const args = {};
            if (delegator)
                args._delegator = delegator;
            if (serviceProvider)
                args._serviceProvider = serviceProvider;
            if (increaseAmount)
                args._increaseAmount = increaseAmount;
            return this.client.getContractEvents({
                address: this.address,
                abi,
                eventName: 'IncreaseDelegatedStake',
                fromBlock,
                args
            });
        };
        this.getUndelegateStakeRequestedEvents = ({ fromBlock = BigInt(0), delegator, serviceProvider }) => {
            const args = {};
            if (delegator)
                args._delegator = delegator;
            if (serviceProvider)
                args._serviceProvider = serviceProvider;
            return this.client.getContractEvents({
                address: this.address,
                abi,
                eventName: 'UndelegateStakeRequested',
                fromBlock,
                args
            });
        };
        this.getUndelegateStakeRequestEvaluatedEvents = ({ fromBlock = BigInt(0), delegator, serviceProvider }) => {
            const args = {};
            if (delegator)
                args._delegator = delegator;
            if (serviceProvider)
                args._serviceProvider = serviceProvider;
            return this.client.getContractEvents({
                address: this.address,
                abi,
                eventName: 'UndelegateStakeRequestEvaluated',
                fromBlock,
                args
            });
        };
        this.getUndelegateStakeRequestCancelledEvents = ({ fromBlock = BigInt(0), delegator, serviceProvider }) => {
            const args = {};
            if (delegator)
                args._delegator = delegator;
            if (serviceProvider)
                args._serviceProvider = serviceProvider;
            return this.client.getContractEvents({
                address: this.address,
                abi,
                eventName: 'UndelegateStakeRequestCancelled',
                fromBlock,
                args
            });
        };
        this.getRemoveDelegatorRequestedEvents = ({ fromBlock = BigInt(0), delegator, serviceProvider }) => {
            const args = {};
            if (delegator)
                args._delegator = delegator;
            if (serviceProvider)
                args._serviceProvider = serviceProvider;
            return this.client.getContractEvents({
                address: this.address,
                abi,
                eventName: 'RemoveDelegatorRequested',
                fromBlock,
                args
            });
        };
        this.getRemoveDelegatorEvaluatedEvents = ({ fromBlock = BigInt(0), delegator, serviceProvider, unstakedAmount }) => {
            const args = {};
            if (delegator)
                args._delegator = delegator;
            if (serviceProvider)
                args._serviceProvider = serviceProvider;
            if (unstakedAmount)
                args._unstakedAmount = unstakedAmount;
            return this.client.getContractEvents({
                address: this.address,
                abi,
                eventName: 'RemoveDelegatorRequestEvaluated',
                fromBlock,
                args
            });
        };
        this.getRemoveDelegatorCancelledEvents = ({ fromBlock = BigInt(0), delegator, serviceProvider }) => {
            const args = {};
            if (delegator)
                args._delegator = delegator;
            if (serviceProvider)
                args._serviceProvider = serviceProvider;
            return this.client.getContractEvents({
                address: this.address,
                abi,
                eventName: 'RemoveDelegatorRequestCancelled',
                fromBlock,
                args
            });
        };
        this.getClaimEvents = ({ fromBlock = BigInt(0), claimer }) => {
            const args = {};
            if (claimer)
                args._claimer = claimer;
            return this.client.getContractEvents({
                address: this.address,
                abi,
                eventName: 'Claim',
                fromBlock,
                args
            });
        };
        this.getSlashEvents = ({ fromBlock = BigInt(0), target }) => {
            const args = {};
            if (target)
                args._target = target;
            return this.client.getContractEvents({
                address: this.address,
                abi,
                eventName: 'Slash',
                fromBlock,
                args
            });
        };
        this.getDelegatorsList = ({ serviceProviderAddress }) => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'getDelegatorsList',
            args: [serviceProviderAddress]
        });
        this.getTotalDelegatedToServiceProvider = ({ serviceProviderAddress }) => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'getTotalDelegatedToServiceProvider',
            args: [serviceProviderAddress]
        });
        this.getTotalLockedDelegationForServiceProvider = ({ serviceProviderAddress }) => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'getTotalLockedDelegationForServiceProvider',
            args: [serviceProviderAddress]
        });
        this.getTotalDelegatorStake = ({ delegatorAddress }) => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'getTotalDelegatorStake',
            args: [delegatorAddress]
        });
        this.getDelegatorStakeForServiceProvider = ({ delegatorAddress, serviceProviderAddress }) => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'getDelegatorStakeForServiceProvider',
            args: [delegatorAddress, serviceProviderAddress]
        });
        this.getPendingRemoveDelegatorRequest = ({ delegatorAddress, serviceProviderAddress }) => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'getPendingRemoveDelegatorRequest',
            args: [serviceProviderAddress, delegatorAddress]
        });
        this.getPendingUndelegateRequest = ({ delegatorAddress }) => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'getPendingUndelegateRequest',
            args: [delegatorAddress]
        });
        this.getUndelegateLockupDuration = () => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'getUndelegateLockupDuration'
        });
        this.getMaxDelegators = () => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'getMaxDelegators'
        });
        this.getMinDelegationAmount = () => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'getMinDelegationAmount'
        });
        this.getRemoveDelegatorLockupDuration = () => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'getRemoveDelegatorLockupDuration'
        });
        this.getRemoveDelegatorEvalDuration = () => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'getRemoveDelegatorEvalDuration'
        });
        this.getGovernanceAddress = () => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'getGovernanceAddress'
        });
        this.getServiceProviderFactoryAddress = () => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'getServiceProviderFactoryAddress'
        });
        this.getClaimsManagerAddress = () => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'getClaimsManagerAddress'
        });
        this.getStakingAddress = () => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'getStakingAddress'
        });
        this.getSPMinDelegationAmount = ({ serviceProviderAddress }) => this.client.readContract({
            address: this.address,
            abi,
            functionName: 'getSPMinDelegationAmount',
            args: [serviceProviderAddress]
        });
        this.client = client;
        this.address = address ?? DELEGATE_MANAGER_CONTRACT_ADDRESS;
    }
}
