// @injectivelabs/sdk-ts/dist/cjs/core/modules/distribution/msgs/MsgFundCommunityPool.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const snakecase_keys_1 = __importDefault(require("snakecase-keys"));
const core_proto_ts_1 = require("@injectivelabs/core-proto-ts");
const MsgBase_js_1 = require("../../MsgBase.js");
/**
 * @category Messages
 */
class MsgFundCommunityPool extends MsgBase_js_1.MsgBase {
    static fromJSON(params) {
        return new MsgFundCommunityPool(params);
    }
    toProto() {
        const { params } = this;
        const message = core_proto_ts_1.CosmosDistributionV1Beta1Tx.MsgFundCommunityPool.create();
        message.depositor = params.depositor;
        const coins = params.amount.map((amount) => {
            const coin = core_proto_ts_1.CosmosBaseV1Beta1Coin.Coin.create();
            coin.denom = amount.denom;
            coin.amount = amount.amount;
            return coin;
        });
        message.amount = coins;
        return core_proto_ts_1.CosmosDistributionV1Beta1Tx.MsgFundCommunityPool.fromPartial(message);
    }
    toData() {
        const proto = this.toProto();
        return {
            '@type': '/cosmos.distribution.v1beta1.MsgFundCommunityPool',
            ...proto,
        };
    }
    toAmino() {
        const proto = this.toProto();
        const message = {
            ...(0, snakecase_keys_1.default)(proto),
        };
        return {
            type: 'cosmos-sdk/MsgFundCommunityPool',
            value: message,
        };
    }
    toWeb3Gw() {
        const amino = this.toAmino();
        const { value } = amino;
        return {
            '@type': '/cosmos.distribution.v1beta1.MsgFundCommunityPool',
            ...value,
        };
    }
    toDirectSign() {
        const proto = this.toProto();
        return {
            type: '/cosmos.distribution.v1beta1.MsgFundCommunityPool',
            message: proto,
        };
    }
    toBinary() {
        return core_proto_ts_1.CosmosDistributionV1Beta1Tx.MsgFundCommunityPool.encode(this.toProto()).finish();
    }
}
exports.default = MsgFundCommunityPool;
