// @audius/eth/dist/cjs/staking/staking.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Staking = void 0;
const abi_1 = require("./abi");
const constants_1 = require("./constants");
class Staking {
    constructor(client, { address } = {}) {
        this.token = () => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'token'
        });
        this.totalStaked = () => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'totalStaked'
        });
        this.supportsHistory = () => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'supportsHistory'
        });
        this.totalStakedFor = ({ account }) => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'totalStakedFor',
            args: [account]
        });
        this.totalStakedAt = ({ blockNumber }) => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'totalStakedAt',
            args: [blockNumber]
        });
        this.totalStakedForAt = ({ account, blockNumber }) => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'totalStakedForAt',
            args: [account, blockNumber]
        });
        this.isStaker = ({ account }) => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'isStaker',
            args: [account]
        });
        this.lastClaimedFor = ({ account }) => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'lastClaimedFor',
            args: [account]
        });
        this.getDelegateManagerAddress = () => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getDelegateManagerAddress'
        });
        this.getClaimsManagerAddress = () => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getClaimsManagerAddress'
        });
        this.getServiceProviderFactoryAddress = () => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getServiceProviderFactoryAddress'
        });
        this.getGovernanceAddress = () => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getGovernanceAddress'
        });
        this.client = client;
        this.address = address !== null && address !== void 0 ? address : constants_1.STAKING_CONTRACT_ADDRESS;
    }
}
exports.Staking = Staking;
