// @injectivelabs/utils/dist/esm/test-utils/msgs.js
import { Network, getNetworkInfo, getNetworkEndpoints, } from '@injectivelabs/networks';
import { mockFactory } from './mocks/index.js';
import BigNumber from '../classes/BigNumber.js';
import { DEFAULT_GAS_LIMIT, DEFAULT_GAS_PRICE } from '../constants.js';
export const prepareEip712 = ({ messages, gas = DEFAULT_GAS_LIMIT, network = Network.Mainnet, injectiveAddress = mockFactory.injectiveAddress, ethereumAddress = mockFactory.ethereumAddress, endpoints = {}, accountNumber = 1, sequence = 1, timeoutHeight = 999_999_999, memo = '', }) => {
    const chainInfo = getNetworkInfo(network);
    const actualEndpoints = { ...getNetworkEndpoints(network), ...endpoints };
    const msgs = Array.isArray(messages) ? messages : [messages];
    const web3Msgs = msgs.map((msg) => msg.toWeb3());
    const { tx, eip712 } = mockFactory.eip712Tx({
        ...chainInfo,
        accountNumber,
        sequence,
        timeoutHeight,
        memo,
    });
    const eip712Args = {
        msgs,
        fee: {
            amount: [
                {
                    denom: 'inj',
                    amount: new BigNumber(gas.toString())
                        .times(DEFAULT_GAS_PRICE)
                        .toFixed(),
                },
            ],
            gas: gas.toString(),
            payer: injectiveAddress,
        },
        tx: {
            ...tx,
            sequence: sequence.toString(),
            timeoutHeight: timeoutHeight.toString(),
            accountNumber: accountNumber.toString(),
        },
        evmChainId: eip712.evmChainId,
    };
    const prepareEip712Request = {
        ...eip712,
        chainId: eip712.evmChainId,
        message: web3Msgs,
        gasLimit: gas,
        address: ethereumAddress,
    };
    return { endpoints: actualEndpoints, eip712Args, prepareEip712Request };
};
