// @injectivelabs/sdk-ts/dist/cjs/core/modules/staking/msgs/MsgTransferDelegation.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const snakecase_keys_1 = __importDefault(require("snakecase-keys"));
const core_proto_ts_1 = require("@injectivelabs/core-proto-ts");
const MsgBase_js_1 = require("../../MsgBase.js");
/**
 * @category Messages
 */
class MsgTransferDelegation extends MsgBase_js_1.MsgBase {
    static fromJSON(params) {
        return new MsgTransferDelegation(params);
    }
    toProto() {
        const { params } = this;
        const coinAmount = core_proto_ts_1.CosmosBaseV1Beta1Coin.Coin.create();
        coinAmount.denom = params.amount.denom;
        coinAmount.amount = params.amount.amount;
        const message = core_proto_ts_1.CosmosStakingV1Beta1Tx.MsgTransferDelegation.create();
        message.delegatorAddress = params.injectiveAddress;
        message.validatorAddress = params.validatorAddress;
        message.receiverAddress = params.receiverAddress;
        message.amount = coinAmount;
        return core_proto_ts_1.CosmosStakingV1Beta1Tx.MsgTransferDelegation.fromPartial(message);
    }
    toData() {
        const proto = this.toProto();
        return {
            '@type': '/cosmos.staking.v1beta1.MsgTransferDelegation',
            ...proto,
        };
    }
    toAmino() {
        const proto = this.toProto();
        const message = {
            ...(0, snakecase_keys_1.default)(proto),
        };
        return {
            type: 'cosmos-sdk/MsgTransferDelegation',
            value: message,
        };
    }
    toWeb3Gw() {
        const amino = this.toAmino();
        const { value } = amino;
        return {
            '@type': '/cosmos.staking.v1beta1.MsgTransferDelegation',
            ...value,
        };
    }
    toDirectSign() {
        const proto = this.toProto();
        return {
            type: '/cosmos.staking.v1beta1.MsgTransferDelegation',
            message: proto,
        };
    }
    toBinary() {
        return core_proto_ts_1.CosmosStakingV1Beta1Tx.MsgTransferDelegation.encode(this.toProto()).finish();
    }
}
exports.default = MsgTransferDelegation;
