// @injectivelabs/core-proto-ts/cjs/injective/downtimedetector/v1beta1/genesis.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenesisState = exports.GenesisDowntimeEntry = exports.protobufPackage = void 0;
/* eslint-disable */
var minimal_js_1 = __importDefault(require("protobufjs/minimal.js"));
var timestamp_1 = require("../../../google/protobuf/timestamp.js");
var downtime_duration_1 = require("./downtime_duration.js");
exports.protobufPackage = "injective.downtimedetector.v1beta1";
function createBaseGenesisDowntimeEntry() {
    return { duration: 0, lastDowntime: undefined };
}
exports.GenesisDowntimeEntry = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.duration !== 0) {
            writer.uint32(8).int32(message.duration);
        }
        if (message.lastDowntime !== undefined) {
            timestamp_1.Timestamp.encode(toTimestamp(message.lastDowntime), writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseGenesisDowntimeEntry();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.duration = reader.int32();
                    break;
                case 2:
                    message.lastDowntime = fromTimestamp(timestamp_1.Timestamp.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            duration: isSet(object.duration) ? (0, downtime_duration_1.downtimeFromJSON)(object.duration) : 0,
            lastDowntime: isSet(object.lastDowntime) ? fromJsonTimestamp(object.lastDowntime) : undefined,
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.duration !== undefined && (obj.duration = (0, downtime_duration_1.downtimeToJSON)(message.duration));
        message.lastDowntime !== undefined && (obj.lastDowntime = message.lastDowntime.toISOString());
        return obj;
    },
    create: function (base) {
        return exports.GenesisDowntimeEntry.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b;
        var message = createBaseGenesisDowntimeEntry();
        message.duration = (_a = object.duration) !== null && _a !== void 0 ? _a : 0;
        message.lastDowntime = (_b = object.lastDowntime) !== null && _b !== void 0 ? _b : undefined;
        return message;
    },
};
function createBaseGenesisState() {
    return { downtimes: [], lastBlockTime: undefined };
}
exports.GenesisState = {
    encode: function (message, writer) {
        var e_1, _a;
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        try {
            for (var _b = __values(message.downtimes), _c = _b.next(); !_c.done; _c = _b.next()) {
                var v = _c.value;
                exports.GenesisDowntimeEntry.encode(v, writer.uint32(10).fork()).ldelim();
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        if (message.lastBlockTime !== undefined) {
            timestamp_1.Timestamp.encode(toTimestamp(message.lastBlockTime), writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseGenesisState();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.downtimes.push(exports.GenesisDowntimeEntry.decode(reader, reader.uint32()));
                    break;
                case 2:
                    message.lastBlockTime = fromTimestamp(timestamp_1.Timestamp.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            downtimes: Array.isArray(object === null || object === void 0 ? void 0 : object.downtimes)
                ? object.downtimes.map(function (e) { return exports.GenesisDowntimeEntry.fromJSON(e); })
                : [],
            lastBlockTime: isSet(object.lastBlockTime) ? fromJsonTimestamp(object.lastBlockTime) : undefined,
        };
    },
    toJSON: function (message) {
        var obj = {};
        if (message.downtimes) {
            obj.downtimes = message.downtimes.map(function (e) { return e ? exports.GenesisDowntimeEntry.toJSON(e) : undefined; });
        }
        else {
            obj.downtimes = [];
        }
        message.lastBlockTime !== undefined && (obj.lastBlockTime = message.lastBlockTime.toISOString());
        return obj;
    },
    create: function (base) {
        return exports.GenesisState.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b;
        var message = createBaseGenesisState();
        message.downtimes = ((_a = object.downtimes) === null || _a === void 0 ? void 0 : _a.map(function (e) { return exports.GenesisDowntimeEntry.fromPartial(e); })) || [];
        message.lastBlockTime = (_b = object.lastBlockTime) !== null && _b !== void 0 ? _b : undefined;
        return message;
    },
};
function toTimestamp(date) {
    var seconds = Math.trunc(date.getTime() / 1000).toString();
    var nanos = (date.getTime() % 1000) * 1000000;
    return { seconds: seconds, nanos: nanos };
}
function fromTimestamp(t) {
    var millis = Number(t.seconds) * 1000;
    millis += t.nanos / 1000000;
    return new Date(millis);
}
function fromJsonTimestamp(o) {
    if (o instanceof Date) {
        return o;
    }
    else if (typeof o === "string") {
        return new Date(o);
    }
    else {
        return fromTimestamp(timestamp_1.Timestamp.fromJSON(o));
    }
}
function isSet(value) {
    return value !== null && value !== undefined;
}
