// @audius/spl/dist/cjs/reward-manager/AttestationLayout.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.attestationLayout = exports.AttestationLayout = void 0;
const buffer_layout_1 = require("@solana/buffer-layout");
const buffer_layout_utils_1 = require("@solana/buffer-layout-utils");
const layout_utils_1 = require("../layout-utils");
class AttestationLayout extends buffer_layout_1.Layout {
    constructor(property) {
        super(
        // ethAddress().span + 1 + u64().span + 1 + 32 + 1 + ethAddress().span,
        // 20 + 1 + 8 + 1 + 32 + 1 + 20 = 83
        83, property);
    }
    getSpan(b, offset = 0) {
        return this.span;
    }
    decode(b, offset = 0) {
        const delimiter = Buffer.from('_', 'utf-8');
        const recipientEthAddress = (0, layout_utils_1.ethAddress)().decode(b, offset);
        offset += (0, layout_utils_1.ethAddress)().span + 1;
        const amount = (0, buffer_layout_utils_1.u64)().decode(b, offset);
        offset += (0, buffer_layout_utils_1.u64)().span + 1;
        const delimiterIndex = b
            .slice(offset)
            .findIndex((v) => v === delimiter[0] || v === 0);
        const disbursementSpan = delimiterIndex > -1 ? delimiterIndex : b.byteLength - offset;
        const disbursementIdBlob = (0, buffer_layout_1.blob)(disbursementSpan).decode(b, offset);
        const disbursementId = Buffer.from(disbursementIdBlob).toString('utf-8');
        offset += disbursementSpan + 1;
        return {
            recipientEthAddress,
            amount,
            disbursementId,
            antiAbuseOracleEthAddress: offset < b.byteLength ? (0, layout_utils_1.ethAddress)().decode(b, offset) : null
        };
    }
    encode(src, b, offset = 0) {
        const delimiter = Buffer.from('_', 'utf-8');
        let layoutOffset = offset;
        layoutOffset += (0, layout_utils_1.ethAddress)().encode(src.recipientEthAddress, b, layoutOffset);
        layoutOffset += (0, buffer_layout_1.blob)(1).encode(delimiter, b, layoutOffset);
        layoutOffset += (0, buffer_layout_utils_1.u64)().encode(src.amount, b, layoutOffset);
        layoutOffset += (0, buffer_layout_1.blob)(1).encode(delimiter, b, layoutOffset);
        layoutOffset += (0, buffer_layout_1.utf8)(32).encode(src.disbursementId, b, layoutOffset);
        if (src.antiAbuseOracleEthAddress) {
            layoutOffset += (0, buffer_layout_1.blob)(1).encode(delimiter, b, layoutOffset);
            layoutOffset += (0, layout_utils_1.ethAddress)().encode(src.antiAbuseOracleEthAddress, b, layoutOffset);
        }
        return layoutOffset - offset;
    }
}
exports.AttestationLayout = AttestationLayout;
const attestationLayout = (property) => new AttestationLayout(property);
exports.attestationLayout = attestationLayout;
