// @audius/spl/dist/cjs/reward-manager/RewardManagerProgram.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RewardManagerProgram = void 0;
const buffer_layout_1 = require("@solana/buffer-layout");
const buffer_layout_utils_1 = require("@solana/buffer-layout-utils");
const spl_token_1 = require("@solana/spl-token");
const web3_js_1 = require("@solana/web3.js");
const layout_utils_1 = require("../layout-utils");
const AttestationLayout_1 = require("./AttestationLayout");
const SenderAttestationLayout_1 = require("./SenderAttestationLayout");
const constants_1 = require("./constants");
const encoder = new TextEncoder();
const SENDER_SEED_PREFIX = 'S_';
const SENDER_SEED_PREFIX_BYTES = encoder.encode(SENDER_SEED_PREFIX);
const ATTESTATIONS_SEED_PREFIX = 'V_';
const ATTESTATIONS_SEED_PREFIX_BYTES = encoder.encode(ATTESTATIONS_SEED_PREFIX);
const DISBURSEMENT_SEED_PREFIX = 'T_';
const DISBURSEMENT_SEED_PREFIX_BYTES = encoder.encode(DISBURSEMENT_SEED_PREFIX);
class RewardManagerProgram {
    static createSenderInstruction({ senderEthAddress, operatorEthAddress, rewardManagerState, manager, authority, payer, sender, rewardManagerProgramId = RewardManagerProgram.programId }) {
        const data = Buffer.alloc(RewardManagerProgram.layouts.createSenderInstructionData.span);
        RewardManagerProgram.layouts.createSenderInstructionData.encode({
            instruction: constants_1.RewardManagerInstruction.CreateSender,
            senderEthAddress,
            operatorEthAddress
        }, data);
        const keys = [
            { pubkey: rewardManagerState, isSigner: false, isWritable: false },
            { pubkey: manager, isSigner: true, isWritable: false },
            { pubkey: authority, isSigner: false, isWritable: false },
            { pubkey: payer, isSigner: true, isWritable: false },
            { pubkey: sender, isSigner: false, isWritable: true },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
            { pubkey: web3_js_1.SYSVAR_RENT_PUBKEY, isSigner: false, isWritable: false }
        ];
        return new web3_js_1.TransactionInstruction({
            programId: rewardManagerProgramId,
            keys,
            data
        });
    }
    static decodeCreateSenderInstruction({ programId, keys: [rewardManagerState, manager, authority, payer, sender, systemProgramId, rent], data }) {
        return {
            programId,
            keys: {
                rewardManagerState,
                manager,
                authority,
                payer,
                sender,
                systemProgramId,
                rent
            },
            data: RewardManagerProgram.layouts.createSenderInstructionData.decode(data)
        };
    }
    static createSenderPublicInstruction({ senderEthAddress, operatorEthAddress, rewardManagerState, authority, payer, sender, existingSenders, rewardManagerProgramId }) {
        const data = Buffer.alloc(RewardManagerProgram.layouts.createSenderPublicInstructionData.span);
        RewardManagerProgram.layouts.createSenderPublicInstructionData.encode({
            instruction: constants_1.RewardManagerInstruction.CreateSenderPublic,
            senderEthAddress,
            operatorEthAddress
        }, data);
        const keys = [
            { pubkey: rewardManagerState, isSigner: false, isWritable: false },
            { pubkey: authority, isSigner: false, isWritable: false },
            { pubkey: payer, isSigner: true, isWritable: true },
            { pubkey: sender, isSigner: false, isWritable: true },
            {
                pubkey: web3_js_1.SYSVAR_INSTRUCTIONS_PUBKEY,
                isSigner: false,
                isWritable: false
            },
            { pubkey: web3_js_1.SYSVAR_RENT_PUBKEY, isSigner: false, isWritable: false },
            { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false },
            ...existingSenders.map((pubkey) => ({
                pubkey,
                isSigner: false,
                isWritable: false
            }))
        ];
        return new web3_js_1.TransactionInstruction({
            programId: rewardManagerProgramId,
            keys,
            data
        });
    }
    static decodeCreateSenderPublicInstruction({ programId, keys: [rewardManagerState, authority, payer, sender, sysvarInstructions, rent, systemProgramId, ...existingSenders], data }) {
        return {
            programId,
            keys: {
                rewardManagerState,
                authority,
                payer,
                sender,
                sysvarInstructions,
                rent,
                systemProgramId,
                existingSenders
            },
            data: RewardManagerProgram.layouts.createSenderPublicInstructionData.decode(data)
        };
    }
    static decodeDeleteSenderPublicInstruction({ programId, keys: [rewardManagerState, sender, refunder, sysvarInstructions, ...existingSenders] }) {
        return {
            programId,
            keys: {
                rewardManagerState,
                sender,
                refunder,
                sysvarInstructions,
                existingSenders
            },
            data: {
                instruction: constants_1.RewardManagerInstruction.DeleteSenderPublic
            }
        };
    }
    static decodeSubmitAttestationInstruction({ programId, keys: [attestations, rewardManagerState, authority, payer, sender, rent, sysvarInstructions, systemProgramId], data }) {
        return {
            programId,
            keys: {
                attestations,
                rewardManagerState,
                authority,
                payer,
                sender,
                rent,
                sysvarInstructions,
                systemProgramId
            },
            data: RewardManagerProgram.layouts.submitAttestationInstructionData.decode(data)
        };
    }
    static decodeEvaluateAttestationsInstruction({ programId, keys: [attestations, rewardManagerState, authority, rewardManagerTokenSource, destinationUserbank, disbursementAccount, antiAbuseOracle, payer, rent, tokenProgramId, systemProgramId], data }) {
        return {
            programId,
            keys: {
                attestations,
                rewardManagerState,
                authority,
                rewardManagerTokenSource,
                destinationUserbank,
                disbursementAccount,
                antiAbuseOracle,
                payer,
                rent,
                tokenProgramId,
                systemProgramId
            },
            data: RewardManagerProgram.layouts.evaluateAttestationsInstructionData.decode(data)
        };
    }
    static decodeInstruction(instruction) {
        switch (instruction.data[0]) {
            case constants_1.RewardManagerInstruction.Init:
            case constants_1.RewardManagerInstruction.ChangeManagerAccount:
                throw new Error('Not Implemented');
            case constants_1.RewardManagerInstruction.CreateSender:
                return RewardManagerProgram.decodeCreateSenderInstruction(instruction);
            case constants_1.RewardManagerInstruction.DeleteSender:
                throw new Error('Not Implemented');
            case constants_1.RewardManagerInstruction.CreateSenderPublic:
                return RewardManagerProgram.decodeCreateSenderPublicInstruction(instruction);
            case constants_1.RewardManagerInstruction.DeleteSenderPublic:
                return RewardManagerProgram.decodeDeleteSenderPublicInstruction(instruction);
            case constants_1.RewardManagerInstruction.SubmitAttestation:
                return RewardManagerProgram.decodeSubmitAttestationInstruction(instruction);
            case constants_1.RewardManagerInstruction.EvaluateAttestations:
                return RewardManagerProgram.decodeEvaluateAttestationsInstruction(instruction);
            default:
                throw new Error('Invalid RewardManager Instruction');
        }
    }
    static isCreateSenderInstruction(decoded) {
        return decoded.data.instruction === constants_1.RewardManagerInstruction.CreateSender;
    }
    static isCreateSenderPublicInstruction(decoded) {
        return (decoded.data.instruction === constants_1.RewardManagerInstruction.CreateSenderPublic);
    }
    static isDeleteSenderPublicInstruction(decoded) {
        return (decoded.data.instruction === constants_1.RewardManagerInstruction.DeleteSenderPublic);
    }
    static isSubmitAttestationInstruction(decoded) {
        return (decoded.data.instruction === constants_1.RewardManagerInstruction.SubmitAttestation);
    }
    static isEvaluateAttestationsInstruction(decoded) {
        return (decoded.data.instruction === constants_1.RewardManagerInstruction.EvaluateAttestations);
    }
    static encodeAttestation(attestation) {
        const data = Buffer.alloc((0, AttestationLayout_1.attestationLayout)().span);
        const span = (0, AttestationLayout_1.attestationLayout)().encode(attestation, data);
        return data.subarray(0, span);
    }
    static decodeAttestation(data) {
        return (0, AttestationLayout_1.attestationLayout)().decode(data);
    }
    static decodeAttestationsAccountData(maxAttestations, data) {
        const decoded = this.layouts
            .attestationsAccountData(maxAttestations)
            .decode(data);
        decoded.messages = decoded.messages.slice(0, decoded.count);
        for (let i = 0; i < decoded.messages.length; i++) {
            if (decoded.messages[i].attestation.antiAbuseOracleEthAddress ===
                '0x0000000000000000000000000000000000000000') {
                decoded.messages[i].attestation.antiAbuseOracleEthAddress = null;
            }
        }
        return decoded;
    }
    static deriveAuthority({ programId, rewardManagerState }) {
        return web3_js_1.PublicKey.findProgramAddressSync([rewardManagerState.toBytes().slice(0, 32)], programId)[0];
    }
    static deriveSender({ ethAddress: wallet, programId, authority }) {
        const ethAddressData = (0, layout_utils_1.ethAddress)(wallet);
        const buffer = Buffer.alloc(ethAddressData.span);
        ethAddressData.encode(wallet, buffer);
        const seed = Uint8Array.from([...SENDER_SEED_PREFIX_BYTES, ...buffer]);
        return web3_js_1.PublicKey.findProgramAddressSync([authority.toBytes().slice(0, 32), seed], programId)[0];
    }
    static deriveAttestations({ disbursementId, programId, authority }) {
        const encoder = new TextEncoder();
        const seed = Uint8Array.from([
            ...ATTESTATIONS_SEED_PREFIX_BYTES,
            ...encoder.encode(disbursementId)
        ]);
        return web3_js_1.PublicKey.findProgramAddressSync([authority.toBytes().slice(0, 32), seed], programId)[0];
    }
    static deriveDisbursement({ disbursementId, programId, authority }) {
        const encoder = new TextEncoder();
        const seed = Uint8Array.from([
            ...DISBURSEMENT_SEED_PREFIX_BYTES,
            ...encoder.encode(disbursementId)
        ]);
        return web3_js_1.PublicKey.findProgramAddressSync([authority.toBytes().slice(0, 32), seed], programId)[0];
    }
    static encodeSignature(signature) {
        const recoveryIdString = signature.slice(-2);
        const recoveryIdBuffer = Buffer.from(recoveryIdString, 'hex');
        const strippedSignature = signature
            .substring(0, signature.length - 2)
            .replace('0x', '')
            .padStart(128, '0')
            .substring(0, 128);
        const signatureBuffer = Buffer.from(strippedSignature, 'hex');
        return {
            signature: signatureBuffer,
            recoveryId: recoveryIdBuffer.readInt8()
        };
    }
    static decodeRewardManagerState(accountData) {
        return RewardManagerProgram.layouts.rewardManagerStateData.decode(accountData);
    }
    static encodeSenderAttestation(attestation) {
        const data = Buffer.alloc((0, SenderAttestationLayout_1.senderAttestationLayout)().span);
        const span = (0, SenderAttestationLayout_1.senderAttestationLayout)().encode(attestation, data);
        return data.subarray(0, span);
    }
}
_a = RewardManagerProgram;
RewardManagerProgram.programId = new web3_js_1.PublicKey('DDZDcYdQFEMwcu2Mwo75yGFjJ1mUQyyXLWzhZLEVFcei');
RewardManagerProgram.layouts = {
    createSenderInstructionData: (0, buffer_layout_1.struct)([
        (0, buffer_layout_1.u8)('instruction'),
        (0, layout_utils_1.ethAddress)('senderEthAddress'),
        (0, layout_utils_1.ethAddress)('operatorEthAddress')
    ]),
    createSenderPublicInstructionData: (0, buffer_layout_1.struct)([
        (0, buffer_layout_1.u8)('instruction'),
        (0, layout_utils_1.ethAddress)('senderEthAddress'),
        (0, layout_utils_1.ethAddress)('operatorEthAddress')
    ]),
    evaluateAttestationsInstructionData: (0, buffer_layout_1.struct)([
        (0, buffer_layout_1.u8)('instruction'),
        (0, buffer_layout_utils_1.u64)('amount'),
        (0, layout_utils_1.borshString)(32, 'disbursementId'),
        (0, layout_utils_1.ethAddress)('recipientEthAddress')
    ]),
    submitAttestationInstructionData: (0, buffer_layout_1.struct)([
        (0, buffer_layout_1.u8)('instruction'),
        (0, layout_utils_1.borshString)(32, 'disbursementId')
    ]),
    rewardManagerStateData: (0, buffer_layout_1.struct)([
        (0, buffer_layout_1.u8)('version'),
        (0, buffer_layout_utils_1.publicKey)('tokenAccount'),
        (0, buffer_layout_utils_1.publicKey)('manager'),
        (0, buffer_layout_1.u8)('minVotes')
    ]),
    attestationsAccountData: (maxAttestations) => (0, buffer_layout_1.struct)([
        (0, buffer_layout_1.u8)('version'),
        (0, buffer_layout_utils_1.publicKey)('rewardManagerState'),
        (0, buffer_layout_1.u8)('count'),
        (0, buffer_layout_1.seq)((0, buffer_layout_1.struct)([
            (0, layout_utils_1.ethAddress)('senderEthAddress'),
            (0, AttestationLayout_1.attestationLayout)('attestation'),
            // Though the actual attestation message is only 83 bytes, we allocate
            // 128 bytes for each element of this array on the program side.
            // Thus we add 45 bytes of padding here to be consistent.
            // See: https://github.com/AudiusProject/audius-protocol/blob/dde78ad7e26d9f6fb358fef5d240c5c7e2d25c66/solana-programs/reward-manager/program/src/state/verified_messages.rs#L99
            (0, buffer_layout_1.blob)(45),
            (0, layout_utils_1.ethAddress)('operator')
        ]), maxAttestations, 'messages')
    ])
};
RewardManagerProgram.createSubmitAttestationInstruction = ({ disbursementId, attestations, rewardManagerState, authority, payer, sender, rewardManagerProgramId = RewardManagerProgram.programId }) => {
    const b = Buffer.alloc(_a.layouts.submitAttestationInstructionData.span);
    const length = _a.layouts.submitAttestationInstructionData.encode({
        instruction: constants_1.RewardManagerInstruction.SubmitAttestation,
        disbursementId
    }, b);
    const data = b.subarray(0, length);
    const keys = [
        { pubkey: attestations, isSigner: false, isWritable: true },
        { pubkey: rewardManagerState, isSigner: false, isWritable: false },
        { pubkey: authority, isSigner: false, isWritable: false },
        { pubkey: payer, isSigner: true, isWritable: true },
        { pubkey: sender, isSigner: false, isWritable: false },
        { pubkey: web3_js_1.SYSVAR_RENT_PUBKEY, isSigner: false, isWritable: false },
        {
            pubkey: web3_js_1.SYSVAR_INSTRUCTIONS_PUBKEY,
            isSigner: false,
            isWritable: false
        },
        { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false }
    ];
    return new web3_js_1.TransactionInstruction({
        programId: rewardManagerProgramId,
        keys,
        data
    });
};
RewardManagerProgram.createEvaluateAttestationsInstruction = ({ disbursementId, recipientEthAddress, amount, attestations, rewardManagerState, authority, rewardManagerTokenSource, destinationUserBank, disbursementAccount, antiAbuseOracle, payer, tokenProgramId = spl_token_1.TOKEN_PROGRAM_ID, rewardManagerProgramId = RewardManagerProgram.programId }) => {
    const b = Buffer.alloc(RewardManagerProgram.layouts.evaluateAttestationsInstructionData.span);
    const length = RewardManagerProgram.layouts.evaluateAttestationsInstructionData.encode({
        instruction: constants_1.RewardManagerInstruction.EvaluateAttestations,
        disbursementId,
        amount,
        recipientEthAddress
    }, b);
    const data = b.subarray(0, length);
    const keys = [
        { pubkey: attestations, isSigner: false, isWritable: true },
        { pubkey: rewardManagerState, isSigner: false, isWritable: false },
        { pubkey: authority, isSigner: false, isWritable: false },
        { pubkey: rewardManagerTokenSource, isSigner: false, isWritable: true },
        { pubkey: destinationUserBank, isSigner: false, isWritable: true },
        { pubkey: disbursementAccount, isSigner: false, isWritable: true },
        { pubkey: antiAbuseOracle, isSigner: false, isWritable: false },
        { pubkey: payer, isSigner: true, isWritable: true },
        { pubkey: web3_js_1.SYSVAR_RENT_PUBKEY, isSigner: false, isWritable: false },
        { pubkey: tokenProgramId, isSigner: false, isWritable: false },
        { pubkey: web3_js_1.SystemProgram.programId, isSigner: false, isWritable: false }
    ];
    return new web3_js_1.TransactionInstruction({
        programId: rewardManagerProgramId,
        keys,
        data
    });
};
exports.RewardManagerProgram = RewardManagerProgram;
