// @injectivelabs/core-proto-ts/cjs/injective/evm/v1/chain_config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlobConfig = exports.BlobScheduleConfig = exports.ChainConfig = exports.protobufPackage = void 0;
/* eslint-disable */
var long_1 = __importDefault(require("long"));
var minimal_js_1 = __importDefault(require("protobufjs/minimal.js"));
exports.protobufPackage = "injective.evm.v1";
function createBaseChainConfig() {
    return {
        homesteadBlock: "",
        daoForkBlock: "",
        daoForkSupport: false,
        eip150Block: "",
        eip150Hash: "",
        eip155Block: "",
        eip158Block: "",
        byzantiumBlock: "",
        constantinopleBlock: "",
        petersburgBlock: "",
        istanbulBlock: "",
        muirGlacierBlock: "",
        berlinBlock: "",
        londonBlock: "",
        arrowGlacierBlock: "",
        grayGlacierBlock: "",
        mergeNetsplitBlock: "",
        shanghaiTime: "",
        cancunTime: "",
        pragueTime: "",
        eip155ChainId: "",
        blobScheduleConfig: undefined,
    };
}
exports.ChainConfig = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.homesteadBlock !== "") {
            writer.uint32(10).string(message.homesteadBlock);
        }
        if (message.daoForkBlock !== "") {
            writer.uint32(18).string(message.daoForkBlock);
        }
        if (message.daoForkSupport === true) {
            writer.uint32(24).bool(message.daoForkSupport);
        }
        if (message.eip150Block !== "") {
            writer.uint32(34).string(message.eip150Block);
        }
        if (message.eip150Hash !== "") {
            writer.uint32(42).string(message.eip150Hash);
        }
        if (message.eip155Block !== "") {
            writer.uint32(50).string(message.eip155Block);
        }
        if (message.eip158Block !== "") {
            writer.uint32(58).string(message.eip158Block);
        }
        if (message.byzantiumBlock !== "") {
            writer.uint32(66).string(message.byzantiumBlock);
        }
        if (message.constantinopleBlock !== "") {
            writer.uint32(74).string(message.constantinopleBlock);
        }
        if (message.petersburgBlock !== "") {
            writer.uint32(82).string(message.petersburgBlock);
        }
        if (message.istanbulBlock !== "") {
            writer.uint32(90).string(message.istanbulBlock);
        }
        if (message.muirGlacierBlock !== "") {
            writer.uint32(98).string(message.muirGlacierBlock);
        }
        if (message.berlinBlock !== "") {
            writer.uint32(106).string(message.berlinBlock);
        }
        if (message.londonBlock !== "") {
            writer.uint32(138).string(message.londonBlock);
        }
        if (message.arrowGlacierBlock !== "") {
            writer.uint32(146).string(message.arrowGlacierBlock);
        }
        if (message.grayGlacierBlock !== "") {
            writer.uint32(162).string(message.grayGlacierBlock);
        }
        if (message.mergeNetsplitBlock !== "") {
            writer.uint32(170).string(message.mergeNetsplitBlock);
        }
        if (message.shanghaiTime !== "") {
            writer.uint32(178).string(message.shanghaiTime);
        }
        if (message.cancunTime !== "") {
            writer.uint32(186).string(message.cancunTime);
        }
        if (message.pragueTime !== "") {
            writer.uint32(194).string(message.pragueTime);
        }
        if (message.eip155ChainId !== "") {
            writer.uint32(202).string(message.eip155ChainId);
        }
        if (message.blobScheduleConfig !== undefined) {
            exports.BlobScheduleConfig.encode(message.blobScheduleConfig, writer.uint32(210).fork()).ldelim();
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseChainConfig();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.homesteadBlock = reader.string();
                    break;
                case 2:
                    message.daoForkBlock = reader.string();
                    break;
                case 3:
                    message.daoForkSupport = reader.bool();
                    break;
                case 4:
                    message.eip150Block = reader.string();
                    break;
                case 5:
                    message.eip150Hash = reader.string();
                    break;
                case 6:
                    message.eip155Block = reader.string();
                    break;
                case 7:
                    message.eip158Block = reader.string();
                    break;
                case 8:
                    message.byzantiumBlock = reader.string();
                    break;
                case 9:
                    message.constantinopleBlock = reader.string();
                    break;
                case 10:
                    message.petersburgBlock = reader.string();
                    break;
                case 11:
                    message.istanbulBlock = reader.string();
                    break;
                case 12:
                    message.muirGlacierBlock = reader.string();
                    break;
                case 13:
                    message.berlinBlock = reader.string();
                    break;
                case 17:
                    message.londonBlock = reader.string();
                    break;
                case 18:
                    message.arrowGlacierBlock = reader.string();
                    break;
                case 20:
                    message.grayGlacierBlock = reader.string();
                    break;
                case 21:
                    message.mergeNetsplitBlock = reader.string();
                    break;
                case 22:
                    message.shanghaiTime = reader.string();
                    break;
                case 23:
                    message.cancunTime = reader.string();
                    break;
                case 24:
                    message.pragueTime = reader.string();
                    break;
                case 25:
                    message.eip155ChainId = reader.string();
                    break;
                case 26:
                    message.blobScheduleConfig = exports.BlobScheduleConfig.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            homesteadBlock: isSet(object.homesteadBlock) ? String(object.homesteadBlock) : "",
            daoForkBlock: isSet(object.daoForkBlock) ? String(object.daoForkBlock) : "",
            daoForkSupport: isSet(object.daoForkSupport) ? Boolean(object.daoForkSupport) : false,
            eip150Block: isSet(object.eip150Block) ? String(object.eip150Block) : "",
            eip150Hash: isSet(object.eip150Hash) ? String(object.eip150Hash) : "",
            eip155Block: isSet(object.eip155Block) ? String(object.eip155Block) : "",
            eip158Block: isSet(object.eip158Block) ? String(object.eip158Block) : "",
            byzantiumBlock: isSet(object.byzantiumBlock) ? String(object.byzantiumBlock) : "",
            constantinopleBlock: isSet(object.constantinopleBlock) ? String(object.constantinopleBlock) : "",
            petersburgBlock: isSet(object.petersburgBlock) ? String(object.petersburgBlock) : "",
            istanbulBlock: isSet(object.istanbulBlock) ? String(object.istanbulBlock) : "",
            muirGlacierBlock: isSet(object.muirGlacierBlock) ? String(object.muirGlacierBlock) : "",
            berlinBlock: isSet(object.berlinBlock) ? String(object.berlinBlock) : "",
            londonBlock: isSet(object.londonBlock) ? String(object.londonBlock) : "",
            arrowGlacierBlock: isSet(object.arrowGlacierBlock) ? String(object.arrowGlacierBlock) : "",
            grayGlacierBlock: isSet(object.grayGlacierBlock) ? String(object.grayGlacierBlock) : "",
            mergeNetsplitBlock: isSet(object.mergeNetsplitBlock) ? String(object.mergeNetsplitBlock) : "",
            shanghaiTime: isSet(object.shanghaiTime) ? String(object.shanghaiTime) : "",
            cancunTime: isSet(object.cancunTime) ? String(object.cancunTime) : "",
            pragueTime: isSet(object.pragueTime) ? String(object.pragueTime) : "",
            eip155ChainId: isSet(object.eip155ChainId) ? String(object.eip155ChainId) : "",
            blobScheduleConfig: isSet(object.blobScheduleConfig)
                ? exports.BlobScheduleConfig.fromJSON(object.blobScheduleConfig)
                : undefined,
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.homesteadBlock !== undefined && (obj.homesteadBlock = message.homesteadBlock);
        message.daoForkBlock !== undefined && (obj.daoForkBlock = message.daoForkBlock);
        message.daoForkSupport !== undefined && (obj.daoForkSupport = message.daoForkSupport);
        message.eip150Block !== undefined && (obj.eip150Block = message.eip150Block);
        message.eip150Hash !== undefined && (obj.eip150Hash = message.eip150Hash);
        message.eip155Block !== undefined && (obj.eip155Block = message.eip155Block);
        message.eip158Block !== undefined && (obj.eip158Block = message.eip158Block);
        message.byzantiumBlock !== undefined && (obj.byzantiumBlock = message.byzantiumBlock);
        message.constantinopleBlock !== undefined && (obj.constantinopleBlock = message.constantinopleBlock);
        message.petersburgBlock !== undefined && (obj.petersburgBlock = message.petersburgBlock);
        message.istanbulBlock !== undefined && (obj.istanbulBlock = message.istanbulBlock);
        message.muirGlacierBlock !== undefined && (obj.muirGlacierBlock = message.muirGlacierBlock);
        message.berlinBlock !== undefined && (obj.berlinBlock = message.berlinBlock);
        message.londonBlock !== undefined && (obj.londonBlock = message.londonBlock);
        message.arrowGlacierBlock !== undefined && (obj.arrowGlacierBlock = message.arrowGlacierBlock);
        message.grayGlacierBlock !== undefined && (obj.grayGlacierBlock = message.grayGlacierBlock);
        message.mergeNetsplitBlock !== undefined && (obj.mergeNetsplitBlock = message.mergeNetsplitBlock);
        message.shanghaiTime !== undefined && (obj.shanghaiTime = message.shanghaiTime);
        message.cancunTime !== undefined && (obj.cancunTime = message.cancunTime);
        message.pragueTime !== undefined && (obj.pragueTime = message.pragueTime);
        message.eip155ChainId !== undefined && (obj.eip155ChainId = message.eip155ChainId);
        message.blobScheduleConfig !== undefined && (obj.blobScheduleConfig = message.blobScheduleConfig
            ? exports.BlobScheduleConfig.toJSON(message.blobScheduleConfig)
            : undefined);
        return obj;
    },
    create: function (base) {
        return exports.ChainConfig.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w;
        var message = createBaseChainConfig();
        message.homesteadBlock = (_a = object.homesteadBlock) !== null && _a !== void 0 ? _a : "";
        message.daoForkBlock = (_b = object.daoForkBlock) !== null && _b !== void 0 ? _b : "";
        message.daoForkSupport = (_c = object.daoForkSupport) !== null && _c !== void 0 ? _c : false;
        message.eip150Block = (_d = object.eip150Block) !== null && _d !== void 0 ? _d : "";
        message.eip150Hash = (_e = object.eip150Hash) !== null && _e !== void 0 ? _e : "";
        message.eip155Block = (_f = object.eip155Block) !== null && _f !== void 0 ? _f : "";
        message.eip158Block = (_g = object.eip158Block) !== null && _g !== void 0 ? _g : "";
        message.byzantiumBlock = (_h = object.byzantiumBlock) !== null && _h !== void 0 ? _h : "";
        message.constantinopleBlock = (_j = object.constantinopleBlock) !== null && _j !== void 0 ? _j : "";
        message.petersburgBlock = (_k = object.petersburgBlock) !== null && _k !== void 0 ? _k : "";
        message.istanbulBlock = (_l = object.istanbulBlock) !== null && _l !== void 0 ? _l : "";
        message.muirGlacierBlock = (_m = object.muirGlacierBlock) !== null && _m !== void 0 ? _m : "";
        message.berlinBlock = (_o = object.berlinBlock) !== null && _o !== void 0 ? _o : "";
        message.londonBlock = (_p = object.londonBlock) !== null && _p !== void 0 ? _p : "";
        message.arrowGlacierBlock = (_q = object.arrowGlacierBlock) !== null && _q !== void 0 ? _q : "";
        message.grayGlacierBlock = (_r = object.grayGlacierBlock) !== null && _r !== void 0 ? _r : "";
        message.mergeNetsplitBlock = (_s = object.mergeNetsplitBlock) !== null && _s !== void 0 ? _s : "";
        message.shanghaiTime = (_t = object.shanghaiTime) !== null && _t !== void 0 ? _t : "";
        message.cancunTime = (_u = object.cancunTime) !== null && _u !== void 0 ? _u : "";
        message.pragueTime = (_v = object.pragueTime) !== null && _v !== void 0 ? _v : "";
        message.eip155ChainId = (_w = object.eip155ChainId) !== null && _w !== void 0 ? _w : "";
        message.blobScheduleConfig = (object.blobScheduleConfig !== undefined && object.blobScheduleConfig !== null)
            ? exports.BlobScheduleConfig.fromPartial(object.blobScheduleConfig)
            : undefined;
        return message;
    },
};
function createBaseBlobScheduleConfig() {
    return { cancun: undefined, prague: undefined, osaka: undefined, verkle: undefined };
}
exports.BlobScheduleConfig = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.cancun !== undefined) {
            exports.BlobConfig.encode(message.cancun, writer.uint32(10).fork()).ldelim();
        }
        if (message.prague !== undefined) {
            exports.BlobConfig.encode(message.prague, writer.uint32(18).fork()).ldelim();
        }
        if (message.osaka !== undefined) {
            exports.BlobConfig.encode(message.osaka, writer.uint32(26).fork()).ldelim();
        }
        if (message.verkle !== undefined) {
            exports.BlobConfig.encode(message.verkle, writer.uint32(34).fork()).ldelim();
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseBlobScheduleConfig();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.cancun = exports.BlobConfig.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.prague = exports.BlobConfig.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.osaka = exports.BlobConfig.decode(reader, reader.uint32());
                    break;
                case 4:
                    message.verkle = exports.BlobConfig.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            cancun: isSet(object.cancun) ? exports.BlobConfig.fromJSON(object.cancun) : undefined,
            prague: isSet(object.prague) ? exports.BlobConfig.fromJSON(object.prague) : undefined,
            osaka: isSet(object.osaka) ? exports.BlobConfig.fromJSON(object.osaka) : undefined,
            verkle: isSet(object.verkle) ? exports.BlobConfig.fromJSON(object.verkle) : undefined,
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.cancun !== undefined && (obj.cancun = message.cancun ? exports.BlobConfig.toJSON(message.cancun) : undefined);
        message.prague !== undefined && (obj.prague = message.prague ? exports.BlobConfig.toJSON(message.prague) : undefined);
        message.osaka !== undefined && (obj.osaka = message.osaka ? exports.BlobConfig.toJSON(message.osaka) : undefined);
        message.verkle !== undefined && (obj.verkle = message.verkle ? exports.BlobConfig.toJSON(message.verkle) : undefined);
        return obj;
    },
    create: function (base) {
        return exports.BlobScheduleConfig.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var message = createBaseBlobScheduleConfig();
        message.cancun = (object.cancun !== undefined && object.cancun !== null)
            ? exports.BlobConfig.fromPartial(object.cancun)
            : undefined;
        message.prague = (object.prague !== undefined && object.prague !== null)
            ? exports.BlobConfig.fromPartial(object.prague)
            : undefined;
        message.osaka = (object.osaka !== undefined && object.osaka !== null)
            ? exports.BlobConfig.fromPartial(object.osaka)
            : undefined;
        message.verkle = (object.verkle !== undefined && object.verkle !== null)
            ? exports.BlobConfig.fromPartial(object.verkle)
            : undefined;
        return message;
    },
};
function createBaseBlobConfig() {
    return { target: "0", max: "0", baseFeeUpdateFraction: "0" };
}
exports.BlobConfig = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.target !== "0") {
            writer.uint32(8).uint64(message.target);
        }
        if (message.max !== "0") {
            writer.uint32(16).uint64(message.max);
        }
        if (message.baseFeeUpdateFraction !== "0") {
            writer.uint32(24).uint64(message.baseFeeUpdateFraction);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseBlobConfig();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.target = longToString(reader.uint64());
                    break;
                case 2:
                    message.max = longToString(reader.uint64());
                    break;
                case 3:
                    message.baseFeeUpdateFraction = longToString(reader.uint64());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            target: isSet(object.target) ? String(object.target) : "0",
            max: isSet(object.max) ? String(object.max) : "0",
            baseFeeUpdateFraction: isSet(object.baseFeeUpdateFraction) ? String(object.baseFeeUpdateFraction) : "0",
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.target !== undefined && (obj.target = message.target);
        message.max !== undefined && (obj.max = message.max);
        message.baseFeeUpdateFraction !== undefined && (obj.baseFeeUpdateFraction = message.baseFeeUpdateFraction);
        return obj;
    },
    create: function (base) {
        return exports.BlobConfig.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b, _c;
        var message = createBaseBlobConfig();
        message.target = (_a = object.target) !== null && _a !== void 0 ? _a : "0";
        message.max = (_b = object.max) !== null && _b !== void 0 ? _b : "0";
        message.baseFeeUpdateFraction = (_c = object.baseFeeUpdateFraction) !== null && _c !== void 0 ? _c : "0";
        return message;
    },
};
function longToString(long) {
    return long.toString();
}
if (minimal_js_1.default.util.Long !== long_1.default) {
    minimal_js_1.default.util.Long = long_1.default;
    minimal_js_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
