// @injectivelabs/core-proto-ts/cjs/injective/evm/v1/events.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventBlockBloom = exports.EventMessage = exports.EventTxLog = exports.EventEthereumTx = exports.protobufPackage = void 0;
/* eslint-disable */
var minimal_js_1 = __importDefault(require("protobufjs/minimal.js"));
exports.protobufPackage = "injective.evm.v1";
function createBaseEventEthereumTx() {
    return { amount: "", ethHash: "", index: "", gasUsed: "", hash: "", recipient: "", ethTxFailed: "" };
}
exports.EventEthereumTx = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.amount !== "") {
            writer.uint32(10).string(message.amount);
        }
        if (message.ethHash !== "") {
            writer.uint32(18).string(message.ethHash);
        }
        if (message.index !== "") {
            writer.uint32(26).string(message.index);
        }
        if (message.gasUsed !== "") {
            writer.uint32(34).string(message.gasUsed);
        }
        if (message.hash !== "") {
            writer.uint32(42).string(message.hash);
        }
        if (message.recipient !== "") {
            writer.uint32(50).string(message.recipient);
        }
        if (message.ethTxFailed !== "") {
            writer.uint32(58).string(message.ethTxFailed);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseEventEthereumTx();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.amount = reader.string();
                    break;
                case 2:
                    message.ethHash = reader.string();
                    break;
                case 3:
                    message.index = reader.string();
                    break;
                case 4:
                    message.gasUsed = reader.string();
                    break;
                case 5:
                    message.hash = reader.string();
                    break;
                case 6:
                    message.recipient = reader.string();
                    break;
                case 7:
                    message.ethTxFailed = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            amount: isSet(object.amount) ? String(object.amount) : "",
            ethHash: isSet(object.ethHash) ? String(object.ethHash) : "",
            index: isSet(object.index) ? String(object.index) : "",
            gasUsed: isSet(object.gasUsed) ? String(object.gasUsed) : "",
            hash: isSet(object.hash) ? String(object.hash) : "",
            recipient: isSet(object.recipient) ? String(object.recipient) : "",
            ethTxFailed: isSet(object.ethTxFailed) ? String(object.ethTxFailed) : "",
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.amount !== undefined && (obj.amount = message.amount);
        message.ethHash !== undefined && (obj.ethHash = message.ethHash);
        message.index !== undefined && (obj.index = message.index);
        message.gasUsed !== undefined && (obj.gasUsed = message.gasUsed);
        message.hash !== undefined && (obj.hash = message.hash);
        message.recipient !== undefined && (obj.recipient = message.recipient);
        message.ethTxFailed !== undefined && (obj.ethTxFailed = message.ethTxFailed);
        return obj;
    },
    create: function (base) {
        return exports.EventEthereumTx.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b, _c, _d, _e, _f, _g;
        var message = createBaseEventEthereumTx();
        message.amount = (_a = object.amount) !== null && _a !== void 0 ? _a : "";
        message.ethHash = (_b = object.ethHash) !== null && _b !== void 0 ? _b : "";
        message.index = (_c = object.index) !== null && _c !== void 0 ? _c : "";
        message.gasUsed = (_d = object.gasUsed) !== null && _d !== void 0 ? _d : "";
        message.hash = (_e = object.hash) !== null && _e !== void 0 ? _e : "";
        message.recipient = (_f = object.recipient) !== null && _f !== void 0 ? _f : "";
        message.ethTxFailed = (_g = object.ethTxFailed) !== null && _g !== void 0 ? _g : "";
        return message;
    },
};
function createBaseEventTxLog() {
    return { txLogs: [] };
}
exports.EventTxLog = {
    encode: function (message, writer) {
        var e_1, _a;
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        try {
            for (var _b = __values(message.txLogs), _c = _b.next(); !_c.done; _c = _b.next()) {
                var v = _c.value;
                writer.uint32(10).string(v);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseEventTxLog();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.txLogs.push(reader.string());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return { txLogs: Array.isArray(object === null || object === void 0 ? void 0 : object.txLogs) ? object.txLogs.map(function (e) { return String(e); }) : [] };
    },
    toJSON: function (message) {
        var obj = {};
        if (message.txLogs) {
            obj.txLogs = message.txLogs.map(function (e) { return e; });
        }
        else {
            obj.txLogs = [];
        }
        return obj;
    },
    create: function (base) {
        return exports.EventTxLog.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a;
        var message = createBaseEventTxLog();
        message.txLogs = ((_a = object.txLogs) === null || _a === void 0 ? void 0 : _a.map(function (e) { return e; })) || [];
        return message;
    },
};
function createBaseEventMessage() {
    return { module: "", sender: "", txType: "" };
}
exports.EventMessage = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.module !== "") {
            writer.uint32(10).string(message.module);
        }
        if (message.sender !== "") {
            writer.uint32(18).string(message.sender);
        }
        if (message.txType !== "") {
            writer.uint32(26).string(message.txType);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseEventMessage();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.module = reader.string();
                    break;
                case 2:
                    message.sender = reader.string();
                    break;
                case 3:
                    message.txType = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            module: isSet(object.module) ? String(object.module) : "",
            sender: isSet(object.sender) ? String(object.sender) : "",
            txType: isSet(object.txType) ? String(object.txType) : "",
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.module !== undefined && (obj.module = message.module);
        message.sender !== undefined && (obj.sender = message.sender);
        message.txType !== undefined && (obj.txType = message.txType);
        return obj;
    },
    create: function (base) {
        return exports.EventMessage.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b, _c;
        var message = createBaseEventMessage();
        message.module = (_a = object.module) !== null && _a !== void 0 ? _a : "";
        message.sender = (_b = object.sender) !== null && _b !== void 0 ? _b : "";
        message.txType = (_c = object.txType) !== null && _c !== void 0 ? _c : "";
        return message;
    },
};
function createBaseEventBlockBloom() {
    return { bloom: "" };
}
exports.EventBlockBloom = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.bloom !== "") {
            writer.uint32(10).string(message.bloom);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseEventBlockBloom();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.bloom = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return { bloom: isSet(object.bloom) ? String(object.bloom) : "" };
    },
    toJSON: function (message) {
        var obj = {};
        message.bloom !== undefined && (obj.bloom = message.bloom);
        return obj;
    },
    create: function (base) {
        return exports.EventBlockBloom.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a;
        var message = createBaseEventBlockBloom();
        message.bloom = (_a = object.bloom) !== null && _a !== void 0 ? _a : "";
        return message;
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
