// @injectivelabs/core-proto-ts/cjs/injective/evm/v1/params.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = exports.protobufPackage = void 0;
/* eslint-disable */
var long_1 = __importDefault(require("long"));
var minimal_js_1 = __importDefault(require("protobufjs/minimal.js"));
var chain_config_1 = require("./chain_config.js");
exports.protobufPackage = "injective.evm.v1";
function createBaseParams() {
    return {
        evmDenom: "",
        enableCreate: false,
        enableCall: false,
        extraEips: [],
        chainConfig: undefined,
        allowUnprotectedTxs: false,
        authorizedDeployers: [],
        permissioned: false,
    };
}
exports.Params = {
    encode: function (message, writer) {
        var e_1, _a, e_2, _b;
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.evmDenom !== "") {
            writer.uint32(10).string(message.evmDenom);
        }
        if (message.enableCreate === true) {
            writer.uint32(16).bool(message.enableCreate);
        }
        if (message.enableCall === true) {
            writer.uint32(24).bool(message.enableCall);
        }
        writer.uint32(34).fork();
        try {
            for (var _c = __values(message.extraEips), _d = _c.next(); !_d.done; _d = _c.next()) {
                var v = _d.value;
                writer.int64(v);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
            }
            finally { if (e_1) throw e_1.error; }
        }
        writer.ldelim();
        if (message.chainConfig !== undefined) {
            chain_config_1.ChainConfig.encode(message.chainConfig, writer.uint32(42).fork()).ldelim();
        }
        if (message.allowUnprotectedTxs === true) {
            writer.uint32(48).bool(message.allowUnprotectedTxs);
        }
        try {
            for (var _e = __values(message.authorizedDeployers), _f = _e.next(); !_f.done; _f = _e.next()) {
                var v = _f.value;
                writer.uint32(58).string(v);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
            }
            finally { if (e_2) throw e_2.error; }
        }
        if (message.permissioned === true) {
            writer.uint32(64).bool(message.permissioned);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseParams();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.evmDenom = reader.string();
                    break;
                case 2:
                    message.enableCreate = reader.bool();
                    break;
                case 3:
                    message.enableCall = reader.bool();
                    break;
                case 4:
                    if ((tag & 7) === 2) {
                        var end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.extraEips.push(longToString(reader.int64()));
                        }
                    }
                    else {
                        message.extraEips.push(longToString(reader.int64()));
                    }
                    break;
                case 5:
                    message.chainConfig = chain_config_1.ChainConfig.decode(reader, reader.uint32());
                    break;
                case 6:
                    message.allowUnprotectedTxs = reader.bool();
                    break;
                case 7:
                    message.authorizedDeployers.push(reader.string());
                    break;
                case 8:
                    message.permissioned = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            evmDenom: isSet(object.evmDenom) ? String(object.evmDenom) : "",
            enableCreate: isSet(object.enableCreate) ? Boolean(object.enableCreate) : false,
            enableCall: isSet(object.enableCall) ? Boolean(object.enableCall) : false,
            extraEips: Array.isArray(object === null || object === void 0 ? void 0 : object.extraEips) ? object.extraEips.map(function (e) { return String(e); }) : [],
            chainConfig: isSet(object.chainConfig) ? chain_config_1.ChainConfig.fromJSON(object.chainConfig) : undefined,
            allowUnprotectedTxs: isSet(object.allowUnprotectedTxs) ? Boolean(object.allowUnprotectedTxs) : false,
            authorizedDeployers: Array.isArray(object === null || object === void 0 ? void 0 : object.authorizedDeployers)
                ? object.authorizedDeployers.map(function (e) { return String(e); })
                : [],
            permissioned: isSet(object.permissioned) ? Boolean(object.permissioned) : false,
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.evmDenom !== undefined && (obj.evmDenom = message.evmDenom);
        message.enableCreate !== undefined && (obj.enableCreate = message.enableCreate);
        message.enableCall !== undefined && (obj.enableCall = message.enableCall);
        if (message.extraEips) {
            obj.extraEips = message.extraEips.map(function (e) { return e; });
        }
        else {
            obj.extraEips = [];
        }
        message.chainConfig !== undefined &&
            (obj.chainConfig = message.chainConfig ? chain_config_1.ChainConfig.toJSON(message.chainConfig) : undefined);
        message.allowUnprotectedTxs !== undefined && (obj.allowUnprotectedTxs = message.allowUnprotectedTxs);
        if (message.authorizedDeployers) {
            obj.authorizedDeployers = message.authorizedDeployers.map(function (e) { return e; });
        }
        else {
            obj.authorizedDeployers = [];
        }
        message.permissioned !== undefined && (obj.permissioned = message.permissioned);
        return obj;
    },
    create: function (base) {
        return exports.Params.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b, _c, _d, _e, _f, _g;
        var message = createBaseParams();
        message.evmDenom = (_a = object.evmDenom) !== null && _a !== void 0 ? _a : "";
        message.enableCreate = (_b = object.enableCreate) !== null && _b !== void 0 ? _b : false;
        message.enableCall = (_c = object.enableCall) !== null && _c !== void 0 ? _c : false;
        message.extraEips = ((_d = object.extraEips) === null || _d === void 0 ? void 0 : _d.map(function (e) { return e; })) || [];
        message.chainConfig = (object.chainConfig !== undefined && object.chainConfig !== null)
            ? chain_config_1.ChainConfig.fromPartial(object.chainConfig)
            : undefined;
        message.allowUnprotectedTxs = (_e = object.allowUnprotectedTxs) !== null && _e !== void 0 ? _e : false;
        message.authorizedDeployers = ((_f = object.authorizedDeployers) === null || _f === void 0 ? void 0 : _f.map(function (e) { return e; })) || [];
        message.permissioned = (_g = object.permissioned) !== null && _g !== void 0 ? _g : false;
        return message;
    },
};
function longToString(long) {
    return long.toString();
}
if (minimal_js_1.default.util.Long !== long_1.default) {
    minimal_js_1.default.util.Long = long_1.default;
    minimal_js_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
