// @injectivelabs/core-proto-ts/cjs/injective/evm/v1/trace_config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TraceConfig = exports.protobufPackage = void 0;
/* eslint-disable */
var long_1 = __importDefault(require("long"));
var minimal_js_1 = __importDefault(require("protobufjs/minimal.js"));
var chain_config_1 = require("./chain_config.js");
exports.protobufPackage = "injective.evm.v1";
function createBaseTraceConfig() {
    return {
        tracer: "",
        timeout: "",
        reexec: "0",
        disableStack: false,
        disableStorage: false,
        debug: false,
        limit: 0,
        overrides: undefined,
        enableMemory: false,
        enableReturnData: false,
        tracerJsonConfig: "",
        stateOverrides: new Uint8Array(),
        blockOverrides: new Uint8Array(),
    };
}
exports.TraceConfig = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.tracer !== "") {
            writer.uint32(10).string(message.tracer);
        }
        if (message.timeout !== "") {
            writer.uint32(18).string(message.timeout);
        }
        if (message.reexec !== "0") {
            writer.uint32(24).uint64(message.reexec);
        }
        if (message.disableStack === true) {
            writer.uint32(40).bool(message.disableStack);
        }
        if (message.disableStorage === true) {
            writer.uint32(48).bool(message.disableStorage);
        }
        if (message.debug === true) {
            writer.uint32(64).bool(message.debug);
        }
        if (message.limit !== 0) {
            writer.uint32(72).int32(message.limit);
        }
        if (message.overrides !== undefined) {
            chain_config_1.ChainConfig.encode(message.overrides, writer.uint32(82).fork()).ldelim();
        }
        if (message.enableMemory === true) {
            writer.uint32(88).bool(message.enableMemory);
        }
        if (message.enableReturnData === true) {
            writer.uint32(96).bool(message.enableReturnData);
        }
        if (message.tracerJsonConfig !== "") {
            writer.uint32(106).string(message.tracerJsonConfig);
        }
        if (message.stateOverrides.length !== 0) {
            writer.uint32(114).bytes(message.stateOverrides);
        }
        if (message.blockOverrides.length !== 0) {
            writer.uint32(122).bytes(message.blockOverrides);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseTraceConfig();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.tracer = reader.string();
                    break;
                case 2:
                    message.timeout = reader.string();
                    break;
                case 3:
                    message.reexec = longToString(reader.uint64());
                    break;
                case 5:
                    message.disableStack = reader.bool();
                    break;
                case 6:
                    message.disableStorage = reader.bool();
                    break;
                case 8:
                    message.debug = reader.bool();
                    break;
                case 9:
                    message.limit = reader.int32();
                    break;
                case 10:
                    message.overrides = chain_config_1.ChainConfig.decode(reader, reader.uint32());
                    break;
                case 11:
                    message.enableMemory = reader.bool();
                    break;
                case 12:
                    message.enableReturnData = reader.bool();
                    break;
                case 13:
                    message.tracerJsonConfig = reader.string();
                    break;
                case 14:
                    message.stateOverrides = reader.bytes();
                    break;
                case 15:
                    message.blockOverrides = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            tracer: isSet(object.tracer) ? String(object.tracer) : "",
            timeout: isSet(object.timeout) ? String(object.timeout) : "",
            reexec: isSet(object.reexec) ? String(object.reexec) : "0",
            disableStack: isSet(object.disableStack) ? Boolean(object.disableStack) : false,
            disableStorage: isSet(object.disableStorage) ? Boolean(object.disableStorage) : false,
            debug: isSet(object.debug) ? Boolean(object.debug) : false,
            limit: isSet(object.limit) ? Number(object.limit) : 0,
            overrides: isSet(object.overrides) ? chain_config_1.ChainConfig.fromJSON(object.overrides) : undefined,
            enableMemory: isSet(object.enableMemory) ? Boolean(object.enableMemory) : false,
            enableReturnData: isSet(object.enableReturnData) ? Boolean(object.enableReturnData) : false,
            tracerJsonConfig: isSet(object.tracerJsonConfig) ? String(object.tracerJsonConfig) : "",
            stateOverrides: isSet(object.stateOverrides) ? bytesFromBase64(object.stateOverrides) : new Uint8Array(),
            blockOverrides: isSet(object.blockOverrides) ? bytesFromBase64(object.blockOverrides) : new Uint8Array(),
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.tracer !== undefined && (obj.tracer = message.tracer);
        message.timeout !== undefined && (obj.timeout = message.timeout);
        message.reexec !== undefined && (obj.reexec = message.reexec);
        message.disableStack !== undefined && (obj.disableStack = message.disableStack);
        message.disableStorage !== undefined && (obj.disableStorage = message.disableStorage);
        message.debug !== undefined && (obj.debug = message.debug);
        message.limit !== undefined && (obj.limit = Math.round(message.limit));
        message.overrides !== undefined &&
            (obj.overrides = message.overrides ? chain_config_1.ChainConfig.toJSON(message.overrides) : undefined);
        message.enableMemory !== undefined && (obj.enableMemory = message.enableMemory);
        message.enableReturnData !== undefined && (obj.enableReturnData = message.enableReturnData);
        message.tracerJsonConfig !== undefined && (obj.tracerJsonConfig = message.tracerJsonConfig);
        message.stateOverrides !== undefined &&
            (obj.stateOverrides = base64FromBytes(message.stateOverrides !== undefined ? message.stateOverrides : new Uint8Array()));
        message.blockOverrides !== undefined &&
            (obj.blockOverrides = base64FromBytes(message.blockOverrides !== undefined ? message.blockOverrides : new Uint8Array()));
        return obj;
    },
    create: function (base) {
        return exports.TraceConfig.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
        var message = createBaseTraceConfig();
        message.tracer = (_a = object.tracer) !== null && _a !== void 0 ? _a : "";
        message.timeout = (_b = object.timeout) !== null && _b !== void 0 ? _b : "";
        message.reexec = (_c = object.reexec) !== null && _c !== void 0 ? _c : "0";
        message.disableStack = (_d = object.disableStack) !== null && _d !== void 0 ? _d : false;
        message.disableStorage = (_e = object.disableStorage) !== null && _e !== void 0 ? _e : false;
        message.debug = (_f = object.debug) !== null && _f !== void 0 ? _f : false;
        message.limit = (_g = object.limit) !== null && _g !== void 0 ? _g : 0;
        message.overrides = (object.overrides !== undefined && object.overrides !== null)
            ? chain_config_1.ChainConfig.fromPartial(object.overrides)
            : undefined;
        message.enableMemory = (_h = object.enableMemory) !== null && _h !== void 0 ? _h : false;
        message.enableReturnData = (_j = object.enableReturnData) !== null && _j !== void 0 ? _j : false;
        message.tracerJsonConfig = (_k = object.tracerJsonConfig) !== null && _k !== void 0 ? _k : "";
        message.stateOverrides = (_l = object.stateOverrides) !== null && _l !== void 0 ? _l : new Uint8Array();
        message.blockOverrides = (_m = object.blockOverrides) !== null && _m !== void 0 ? _m : new Uint8Array();
        return message;
    },
};
var tsProtoGlobalThis = (function () {
    if (typeof globalThis !== "undefined") {
        return globalThis;
    }
    if (typeof self !== "undefined") {
        return self;
    }
    if (typeof window !== "undefined") {
        return window;
    }
    if (typeof global !== "undefined") {
        return global;
    }
    throw "Unable to locate global object";
})();
function bytesFromBase64(b64) {
    if (tsProtoGlobalThis.Buffer) {
        return Uint8Array.from(tsProtoGlobalThis.Buffer.from(b64, "base64"));
    }
    else {
        var bin = tsProtoGlobalThis.atob(b64);
        var arr = new Uint8Array(bin.length);
        for (var i = 0; i < bin.length; ++i) {
            arr[i] = bin.charCodeAt(i);
        }
        return arr;
    }
}
function base64FromBytes(arr) {
    if (tsProtoGlobalThis.Buffer) {
        return tsProtoGlobalThis.Buffer.from(arr).toString("base64");
    }
    else {
        var bin_1 = [];
        arr.forEach(function (byte) {
            bin_1.push(String.fromCharCode(byte));
        });
        return tsProtoGlobalThis.btoa(bin_1.join(""));
    }
}
function longToString(long) {
    return long.toString();
}
if (minimal_js_1.default.util.Long !== long_1.default) {
    minimal_js_1.default.util.Long = long_1.default;
    minimal_js_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
