// @injectivelabs/core-proto-ts/cjs/injective/erc20/v1beta1/events.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventDeleteTokenPair = exports.EventCreateTokenPair = exports.protobufPackage = void 0;
/* eslint-disable */
var minimal_js_1 = __importDefault(require("protobufjs/minimal.js"));
exports.protobufPackage = "injective.erc20.v1beta1";
function createBaseEventCreateTokenPair() {
    return { bankDenom: "", erc20Address: "" };
}
exports.EventCreateTokenPair = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.bankDenom !== "") {
            writer.uint32(10).string(message.bankDenom);
        }
        if (message.erc20Address !== "") {
            writer.uint32(18).string(message.erc20Address);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseEventCreateTokenPair();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.bankDenom = reader.string();
                    break;
                case 2:
                    message.erc20Address = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            bankDenom: isSet(object.bankDenom) ? String(object.bankDenom) : "",
            erc20Address: isSet(object.erc20Address) ? String(object.erc20Address) : "",
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.bankDenom !== undefined && (obj.bankDenom = message.bankDenom);
        message.erc20Address !== undefined && (obj.erc20Address = message.erc20Address);
        return obj;
    },
    create: function (base) {
        return exports.EventCreateTokenPair.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b;
        var message = createBaseEventCreateTokenPair();
        message.bankDenom = (_a = object.bankDenom) !== null && _a !== void 0 ? _a : "";
        message.erc20Address = (_b = object.erc20Address) !== null && _b !== void 0 ? _b : "";
        return message;
    },
};
function createBaseEventDeleteTokenPair() {
    return { bankDenom: "" };
}
exports.EventDeleteTokenPair = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.bankDenom !== "") {
            writer.uint32(10).string(message.bankDenom);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseEventDeleteTokenPair();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.bankDenom = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return { bankDenom: isSet(object.bankDenom) ? String(object.bankDenom) : "" };
    },
    toJSON: function (message) {
        var obj = {};
        message.bankDenom !== undefined && (obj.bankDenom = message.bankDenom);
        return obj;
    },
    create: function (base) {
        return exports.EventDeleteTokenPair.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a;
        var message = createBaseEventDeleteTokenPair();
        message.bankDenom = (_a = object.bankDenom) !== null && _a !== void 0 ? _a : "";
        return message;
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
