// @injectivelabs/core-proto-ts/cjs/injective/erc20/v1beta1/params.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = exports.protobufPackage = void 0;
/* eslint-disable */
var minimal_js_1 = __importDefault(require("protobufjs/minimal.js"));
var coin_1 = require("../../../cosmos/base/v1beta1/coin.js");
exports.protobufPackage = "injective.erc20.v1beta1";
function createBaseParams() {
    return { denomCreationFee: undefined };
}
exports.Params = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.denomCreationFee !== undefined) {
            coin_1.Coin.encode(message.denomCreationFee, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseParams();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.denomCreationFee = coin_1.Coin.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return { denomCreationFee: isSet(object.denomCreationFee) ? coin_1.Coin.fromJSON(object.denomCreationFee) : undefined };
    },
    toJSON: function (message) {
        var obj = {};
        message.denomCreationFee !== undefined &&
            (obj.denomCreationFee = message.denomCreationFee ? coin_1.Coin.toJSON(message.denomCreationFee) : undefined);
        return obj;
    },
    create: function (base) {
        return exports.Params.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var message = createBaseParams();
        message.denomCreationFee = (object.denomCreationFee !== undefined && object.denomCreationFee !== null)
            ? coin_1.Coin.fromPartial(object.denomCreationFee)
            : undefined;
        return message;
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
