// @injectivelabs/utils/dist/cjs/classes/BigNumberInBase.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const BigNumber_js_1 = __importDefault(require("./BigNumber.js"));
/**
 * @deprecated - use toHumanReadable instead
 */
class BigNumberInBase extends BigNumber_js_1.default {
    static make(number) {
        return new BigNumberInBase(number);
    }
    minus(n, base) {
        return new BigNumberInBase(super.minus(n, base));
    }
    plus(n, base) {
        return new BigNumberInBase(super.plus(n, base));
    }
    dividedBy(n, base) {
        return new BigNumberInBase(super.dividedBy(n, base));
    }
    div(n, base) {
        return new BigNumberInBase(super.div(n, base));
    }
    multipliedBy(n, base) {
        return new BigNumberInBase(super.multipliedBy(n, base));
    }
    times(n, base) {
        return new BigNumberInBase(super.times(n, base));
    }
    pow(n, base) {
        return new BigNumberInBase(super.pow(n, base));
    }
    toWei(decimals = 18) {
        return new BigNumber_js_1.default(this).multipliedBy(new BigNumber_js_1.default(10).pow(decimals));
    }
}
exports.default = BigNumberInBase;
