// @injectivelabs/utils/dist/cjs/classes/Status.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const enum_js_1 = require("../enum.js");
class Status {
    status;
    constructor(status = enum_js_1.StatusType.Idle) {
        this.status = status;
    }
    get() {
        return this.status;
    }
    set(status) {
        this.status = status;
    }
    is(status) {
        return this.status === status;
    }
    isLoading() {
        return this.is(enum_js_1.StatusType.Loading);
    }
    isNotLoading() {
        return !this.is(enum_js_1.StatusType.Loading);
    }
    isCompleted() {
        return this.is(enum_js_1.StatusType.Completed);
    }
    isConfirmed() {
        return this.is(enum_js_1.StatusType.Confirmed);
    }
    isIdle() {
        return this.is(enum_js_1.StatusType.Idle);
    }
    isError() {
        return this.is(enum_js_1.StatusType.Error);
    }
    setLoading() {
        this.set(enum_js_1.StatusType.Loading);
    }
    setCompleted() {
        this.set(enum_js_1.StatusType.Completed);
    }
    setConfirmed() {
        this.set(enum_js_1.StatusType.Confirmed);
    }
    setError() {
        this.set(enum_js_1.StatusType.Error);
    }
    setIdle() {
        this.set(enum_js_1.StatusType.Idle);
    }
    toggle() {
        this.set(this.status === enum_js_1.StatusType.Idle ? enum_js_1.StatusType.Loading : enum_js_1.StatusType.Idle);
    }
    toString() {
        return this.get();
    }
    valueOf() {
        return this.get();
    }
}
exports.default = Status;
