// @injectivelabs/sdk-ts/dist/cjs/client/chain/grpc/ChainGrpcEvmApi.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainGrpcEvmApi = void 0;
const core_proto_ts_1 = require("@injectivelabs/core-proto-ts");
const exceptions_1 = require("@injectivelabs/exceptions");
const index_js_1 = require("../types/index.js");
const BaseGrpcConsumer_js_1 = __importDefault(require("../../base/BaseGrpcConsumer.js"));
const index_js_2 = require("../transformers/index.js");
/**
 * @category Chain Grpc API
 */
class ChainGrpcEvmApi extends BaseGrpcConsumer_js_1.default {
    module = index_js_1.ChainModule.Evm;
    client;
    constructor(endpoint) {
        super(endpoint);
        this.client = new core_proto_ts_1.InjectiveEvmV1Query.QueryClientImpl(this.getGrpcWebImpl(endpoint));
    }
    async fetchAccount(ethAddress) {
        const request = core_proto_ts_1.InjectiveEvmV1Query.QueryAccountRequest.create();
        request.address = ethAddress;
        try {
            const response = await this.retry(() => this.client.Account(request, this.metadata));
            return index_js_2.ChainGrpcEvmTransformer.accountResponseToAccount(response);
        }
        catch (e) {
            if (e instanceof core_proto_ts_1.InjectiveEvmV1Query.GrpcWebError) {
                throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                    code: (0, exceptions_1.grpcErrorCodeToErrorCode)(e.code),
                    context: 'EvmAccount',
                    contextModule: this.module,
                });
            }
            throw new exceptions_1.GrpcUnaryRequestException(e, {
                code: exceptions_1.UnspecifiedErrorCode,
                context: 'EvmAccount',
                contextModule: this.module,
            });
        }
    }
    async fetchCosmosAccount(ethAddress) {
        const request = core_proto_ts_1.InjectiveEvmV1Query.QueryCosmosAccountRequest.create();
        request.address = ethAddress;
        try {
            const response = await this.retry(() => this.client.CosmosAccount(request, this.metadata));
            return index_js_2.ChainGrpcEvmTransformer.cosmosAccountResponseToCosmosAccount(response);
        }
        catch (e) {
            if (e instanceof core_proto_ts_1.InjectiveEvmV1Query.GrpcWebError) {
                throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                    code: (0, exceptions_1.grpcErrorCodeToErrorCode)(e.code),
                    context: 'CosmosAccount',
                    contextModule: this.module,
                });
            }
            throw new exceptions_1.GrpcUnaryRequestException(e, {
                code: exceptions_1.UnspecifiedErrorCode,
                context: 'CosmosAccount',
                contextModule: this.module,
            });
        }
    }
    async fetchValidatorAccount(consAddress) {
        const request = core_proto_ts_1.InjectiveEvmV1Query.QueryValidatorAccountRequest.create();
        request.consAddress = consAddress;
        try {
            const response = await this.retry(() => this.client.ValidatorAccount(request, this.metadata));
            return index_js_2.ChainGrpcEvmTransformer.validatorAccountResponseToValidatorAccount(response);
        }
        catch (e) {
            if (e instanceof core_proto_ts_1.InjectiveEvmV1Query.GrpcWebError) {
                throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                    code: (0, exceptions_1.grpcErrorCodeToErrorCode)(e.code),
                    context: 'EvmValidatorAccount',
                    contextModule: this.module,
                });
            }
            throw new exceptions_1.GrpcUnaryRequestException(e, {
                code: exceptions_1.UnspecifiedErrorCode,
                context: 'EvmValidatorAccount',
                contextModule: this.module,
            });
        }
    }
    async fetchBalance(ethAddress) {
        const request = core_proto_ts_1.InjectiveEvmV1Query.QueryBalanceRequest.create();
        request.address = ethAddress;
        try {
            const response = await this.retry(() => this.client.Balance(request, this.metadata));
            return response.balance;
        }
        catch (e) {
            if (e instanceof core_proto_ts_1.InjectiveEvmV1Query.GrpcWebError) {
                throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                    code: (0, exceptions_1.grpcErrorCodeToErrorCode)(e.code),
                    context: 'EvmBalance',
                    contextModule: this.module,
                });
            }
            throw new exceptions_1.GrpcUnaryRequestException(e, {
                code: exceptions_1.UnspecifiedErrorCode,
                context: 'EvmBalance',
                contextModule: this.module,
            });
        }
    }
    async fetchStorage(ethAddress, key) {
        const request = core_proto_ts_1.InjectiveEvmV1Query.QueryStorageRequest.create();
        request.address = ethAddress;
        request.key = key;
        try {
            const response = await this.retry(() => this.client.Storage(request, this.metadata));
            return response.value;
        }
        catch (e) {
            if (e instanceof core_proto_ts_1.InjectiveEvmV1Query.GrpcWebError) {
                throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                    code: (0, exceptions_1.grpcErrorCodeToErrorCode)(e.code),
                    context: 'EvmStorage',
                    contextModule: this.module,
                });
            }
            throw new exceptions_1.GrpcUnaryRequestException(e, {
                code: exceptions_1.UnspecifiedErrorCode,
                context: 'EvmStorage',
                contextModule: this.module,
            });
        }
    }
    async fetchCode(ethAddress) {
        const request = core_proto_ts_1.InjectiveEvmV1Query.QueryCodeRequest.create();
        request.address = ethAddress;
        try {
            const response = await this.retry(() => this.client.Code(request, this.metadata));
            return response.code;
        }
        catch (e) {
            if (e instanceof core_proto_ts_1.InjectiveEvmV1Query.GrpcWebError) {
                throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                    code: (0, exceptions_1.grpcErrorCodeToErrorCode)(e.code),
                    context: 'EvmCode',
                    contextModule: this.module,
                });
            }
            throw new exceptions_1.GrpcUnaryRequestException(e, {
                code: exceptions_1.UnspecifiedErrorCode,
                context: 'EvmCode',
                contextModule: this.module,
            });
        }
    }
    async fetchParams() {
        const request = core_proto_ts_1.InjectiveEvmV1Query.QueryParamsRequest.create();
        try {
            const response = await this.retry(() => this.client.Params(request, this.metadata));
            return index_js_2.ChainGrpcEvmTransformer.paramsResponseToParams(response);
        }
        catch (e) {
            if (e instanceof core_proto_ts_1.InjectiveEvmV1Query.GrpcWebError) {
                throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                    code: (0, exceptions_1.grpcErrorCodeToErrorCode)(e.code),
                    context: 'EvmParams',
                    contextModule: this.module,
                });
            }
            throw new exceptions_1.GrpcUnaryRequestException(e, {
                code: exceptions_1.UnspecifiedErrorCode,
                context: 'EvmParams',
                contextModule: this.module,
            });
        }
    }
    async fetchBaseFee() {
        const request = core_proto_ts_1.InjectiveEvmV1Query.QueryBaseFeeRequest.create();
        try {
            const response = await this.retry(() => this.client.BaseFee(request, this.metadata));
            return response.baseFee;
        }
        catch (e) {
            if (e instanceof core_proto_ts_1.InjectiveEvmV1Query.GrpcWebError) {
                throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                    code: (0, exceptions_1.grpcErrorCodeToErrorCode)(e.code),
                    context: 'EvmBaseFee',
                    contextModule: this.module,
                });
            }
            throw new exceptions_1.GrpcUnaryRequestException(e, {
                code: exceptions_1.UnspecifiedErrorCode,
                context: 'EvmBaseFee',
                contextModule: this.module,
            });
        }
    }
}
exports.ChainGrpcEvmApi = ChainGrpcEvmApi;
