// @injectivelabs/sdk-ts/dist/cjs/client/chain/grpc/ChainGrpcTxFeesApi.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainGrpcTxFeesApi = void 0;
const core_proto_ts_1 = require("@injectivelabs/core-proto-ts");
const exceptions_1 = require("@injectivelabs/exceptions");
const index_js_1 = require("../types/index.js");
const BaseGrpcConsumer_js_1 = __importDefault(require("../../base/BaseGrpcConsumer.js"));
const index_js_2 = require("../transformers/index.js");
/**
 * @category Chain Grpc API
 */
class ChainGrpcTxFeesApi extends BaseGrpcConsumer_js_1.default {
    module = index_js_1.ChainModule.TxFees;
    client;
    constructor(endpoint) {
        super(endpoint);
        this.client = new core_proto_ts_1.InjectiveTxFeesV1Beta1Query.QueryClientImpl(this.getGrpcWebImpl(endpoint));
    }
    async fetchModuleParams() {
        const request = core_proto_ts_1.InjectiveTxFeesV1Beta1Query.QueryParamsRequest.create();
        try {
            const response = await this.retry(() => this.client.Params(request, this.metadata));
            return index_js_2.ChainGrpcTxFeesTransformer.moduleParamsResponseToModuleParams(response);
        }
        catch (e) {
            if (e instanceof core_proto_ts_1.InjectiveTxFeesV1Beta1Query.GrpcWebError) {
                throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                    code: (0, exceptions_1.grpcErrorCodeToErrorCode)(e.code),
                    context: 'Params',
                    contextModule: this.module,
                });
            }
            throw new exceptions_1.GrpcUnaryRequestException(e, {
                code: exceptions_1.UnspecifiedErrorCode,
                context: 'Params',
                contextModule: this.module,
            });
        }
    }
    async fetchEipBaseFee() {
        const request = core_proto_ts_1.InjectiveTxFeesV1Beta1Query.QueryEipBaseFeeRequest.create();
        try {
            const response = await this.retry(() => this.client.GetEipBaseFee(request, this.metadata));
            return index_js_2.ChainGrpcTxFeesTransformer.eipBaseFeeResponseToEipBaseFee(response);
        }
        catch (e) {
            if (e instanceof core_proto_ts_1.InjectiveTxFeesV1Beta1Query.GrpcWebError) {
                throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                    code: (0, exceptions_1.grpcErrorCodeToErrorCode)(e.code),
                    context: 'EipBaseFee',
                    contextModule: this.module,
                });
            }
            throw new exceptions_1.GrpcUnaryRequestException(e, {
                code: exceptions_1.UnspecifiedErrorCode,
                context: 'EipBaseFee',
                contextModule: this.module,
            });
        }
    }
}
exports.ChainGrpcTxFeesApi = ChainGrpcTxFeesApi;
