// @injectivelabs/sdk-ts/dist/cjs/client/chain/grpc/ChainGrpcErc20Api.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainGrpcErc20Api = void 0;
const core_proto_ts_1 = require("@injectivelabs/core-proto-ts");
const exceptions_1 = require("@injectivelabs/exceptions");
const index_js_1 = require("../types/index.js");
const BaseGrpcConsumer_js_1 = __importDefault(require("../../base/BaseGrpcConsumer.js"));
const pagination_js_1 = require("../../../utils/pagination.js");
const index_js_2 = require("../transformers/index.js");
const MAX_LIMIT_FOR_SUPPLY = 10000;
/**
 * @category Chain Grpc API
 */
class ChainGrpcErc20Api extends BaseGrpcConsumer_js_1.default {
    module = index_js_1.ChainModule.Erc20;
    client;
    constructor(endpoint) {
        super(endpoint);
        this.client = new core_proto_ts_1.InjectiveErc20V1Beta1Query.QueryClientImpl(this.getGrpcWebImpl(endpoint));
    }
    async fetchModuleParams() {
        const request = core_proto_ts_1.InjectiveErc20V1Beta1Query.QueryParamsRequest.create();
        try {
            const response = await this.retry(() => this.client.Params(request, this.metadata));
            return index_js_2.ChainGrpcErc20Transformer.paramsResponseToParams(response);
        }
        catch (e) {
            if (e instanceof core_proto_ts_1.InjectiveErc20V1Beta1Query.GrpcWebError) {
                throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                    code: (0, exceptions_1.grpcErrorCodeToErrorCode)(e.code),
                    context: 'Erc20Params',
                    contextModule: this.module,
                });
            }
            throw new exceptions_1.GrpcUnaryRequestException(e, {
                code: exceptions_1.UnspecifiedErrorCode,
                context: 'Erc20Params',
                contextModule: this.module,
            });
        }
    }
    async fetchTokenPairs(pagination) {
        const request = core_proto_ts_1.InjectiveErc20V1Beta1Query.QueryAllTokenPairsRequest.create();
        const paginationForRequest = index_js_2.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequest(pagination);
        if (paginationForRequest) {
            request.pagination = paginationForRequest;
        }
        try {
            const response = await this.retry(() => this.client.AllTokenPairs(request, this.metadata));
            return index_js_2.ChainGrpcErc20Transformer.tokenPairsResponseToTokenPairs(response);
        }
        catch (e) {
            if (e instanceof core_proto_ts_1.InjectiveErc20V1Beta1Query.GrpcWebError) {
                throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                    code: (0, exceptions_1.grpcErrorCodeToErrorCode)(e.code),
                    context: 'Erc20TokenPairs',
                    contextModule: this.module,
                });
            }
            throw new exceptions_1.GrpcUnaryRequestException(e, {
                code: exceptions_1.UnspecifiedErrorCode,
                context: 'Erc20TokenPairs',
                contextModule: this.module,
            });
        }
    }
    async fetchAllTokenPairsWithPagination(pagination = { limit: MAX_LIMIT_FOR_SUPPLY }) {
        return (0, pagination_js_1.fetchAllWithPagination)(pagination, this.fetchTokenPairs.bind(this));
    }
    async fetchTokenPairByDenom(denom) {
        const request = core_proto_ts_1.InjectiveErc20V1Beta1Query.QueryTokenPairByDenomRequest.create();
        request.bankDenom = denom;
        try {
            const response = await this.retry(() => this.client.TokenPairByDenom(request, this.metadata));
            if (!response.tokenPair) {
                return undefined;
            }
            return index_js_2.ChainGrpcErc20Transformer.grpcTokenPairToTokenPair(response.tokenPair);
        }
        catch (e) {
            if (e instanceof core_proto_ts_1.InjectiveErc20V1Beta1Query.GrpcWebError) {
                throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                    code: (0, exceptions_1.grpcErrorCodeToErrorCode)(e.code),
                    context: 'Erc20TokenPairByDenom',
                    contextModule: this.module,
                });
            }
            throw new exceptions_1.GrpcUnaryRequestException(e, {
                code: exceptions_1.UnspecifiedErrorCode,
                context: 'Erc20TokenPairByDenom',
                contextModule: this.module,
            });
        }
    }
    async fetchTokenPairByErc20Address(erc20Address) {
        const request = core_proto_ts_1.InjectiveErc20V1Beta1Query.QueryTokenPairByERC20AddressRequest.create();
        request.erc20Address = erc20Address;
        try {
            const response = await this.retry(() => this.client.TokenPairByERC20Address(request, this.metadata));
            if (!response.tokenPair) {
                return undefined;
            }
            return index_js_2.ChainGrpcErc20Transformer.grpcTokenPairToTokenPair(response.tokenPair);
        }
        catch (e) {
            if (e instanceof core_proto_ts_1.InjectiveErc20V1Beta1Query.GrpcWebError) {
                throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                    code: (0, exceptions_1.grpcErrorCodeToErrorCode)(e.code),
                    context: 'Erc20TokenPairByErc20Address',
                    contextModule: this.module,
                });
            }
            throw new exceptions_1.GrpcUnaryRequestException(e, {
                code: exceptions_1.UnspecifiedErrorCode,
                context: 'Erc20TokenPairByErc20Address',
                contextModule: this.module,
            });
        }
    }
}
exports.ChainGrpcErc20Api = ChainGrpcErc20Api;
