// @injectivelabs/sdk-ts/dist/cjs/client/indexer/grpc/IndexerGrpcMegaVaultApi.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndexerGrpcMegaVaultApi = void 0;
const indexer_proto_ts_1 = require("@injectivelabs/indexer-proto-ts");
const exceptions_1 = require("@injectivelabs/exceptions");
const index_js_1 = require("../types/index.js");
const BaseIndexerGrpcConsumer_js_1 = __importDefault(require("../../base/BaseIndexerGrpcConsumer.js"));
const index_js_2 = require("../transformers/index.js");
/**
 * @category Indexer Grpc API
 */
class IndexerGrpcMegaVaultApi extends BaseIndexerGrpcConsumer_js_1.default {
    module = index_js_1.IndexerModule.MegaVault;
    client;
    constructor(endpoint) {
        super(endpoint);
        this.client = new indexer_proto_ts_1.InjectiveMegaVaultRpc.InjectiveMegavaultRPCClientImpl(this.getGrpcWebImpl(endpoint));
    }
    async fetchVault({ vaultAddress }) {
        const request = indexer_proto_ts_1.InjectiveMegaVaultRpc.GetVaultRequest.create();
        request.vaultAddress = vaultAddress;
        try {
            const response = await this.retry(() => this.client.GetVault(request, this.metadata));
            return index_js_2.IndexerGrpcMegaVaultTransformer.vaultResponseToVault(response);
        }
        catch (e) {
            if (e instanceof indexer_proto_ts_1.InjectiveMegaVaultRpc.GrpcWebError) {
                throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                    code: e.code,
                    context: 'fetchVault',
                    contextModule: this.module,
                });
            }
            throw new exceptions_1.GrpcUnaryRequestException(e, {
                code: exceptions_1.UnspecifiedErrorCode,
                context: 'fetchVault',
                contextModule: this.module,
            });
        }
    }
    async fetchVaultUser({ vaultAddress, userAddress, }) {
        const request = indexer_proto_ts_1.InjectiveMegaVaultRpc.GetUserRequest.create();
        request.vaultAddress = vaultAddress;
        request.userAddress = userAddress;
        try {
            const response = await this.retry(() => this.client.GetUser(request, this.metadata));
            return index_js_2.IndexerGrpcMegaVaultTransformer.userResponseToUser(response);
        }
        catch (e) {
            if (e instanceof indexer_proto_ts_1.InjectiveMegaVaultRpc.GrpcWebError) {
                throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                    code: e.code,
                    context: 'fetchVaultUser',
                    contextModule: this.module,
                });
            }
            throw new exceptions_1.GrpcUnaryRequestException(e, {
                code: exceptions_1.UnspecifiedErrorCode,
                context: 'fetchVaultUser',
                contextModule: this.module,
            });
        }
    }
    async fetchVaultSubscriptions({ token, status, perPage, userAddress, vaultAddress, }) {
        const request = indexer_proto_ts_1.InjectiveMegaVaultRpc.ListSubscriptionsRequest.create();
        request.vaultAddress = vaultAddress;
        request.userAddress = userAddress;
        if (status) {
            request.status = status;
        }
        if (perPage) {
            request.perPage = perPage;
        }
        if (token) {
            request.token = token;
        }
        try {
            const response = await this.retry(() => this.client.ListSubscriptions(request, this.metadata));
            return index_js_2.IndexerGrpcMegaVaultTransformer.subscriptionsResponseToSubscriptions(response);
        }
        catch (e) {
            if (e instanceof indexer_proto_ts_1.InjectiveMegaVaultRpc.GrpcWebError) {
                throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                    code: e.code,
                    context: 'fetchVaultSubscriptions',
                    contextModule: this.module,
                });
            }
            throw new exceptions_1.GrpcUnaryRequestException(e, {
                code: exceptions_1.UnspecifiedErrorCode,
                context: 'fetchVaultSubscriptions',
                contextModule: this.module,
            });
        }
    }
    async fetchVaultRedemptions({ token, status, perPage, userAddress, vaultAddress, }) {
        const request = indexer_proto_ts_1.InjectiveMegaVaultRpc.ListRedemptionsRequest.create();
        request.vaultAddress = vaultAddress;
        request.userAddress = userAddress;
        if (status) {
            request.status = status;
        }
        if (perPage) {
            request.perPage = perPage;
        }
        if (token) {
            request.token = token;
        }
        try {
            const response = await this.retry(() => this.client.ListRedemptions(request, this.metadata));
            return index_js_2.IndexerGrpcMegaVaultTransformer.redemptionsResponseToRedemptions(response);
        }
        catch (e) {
            if (e instanceof indexer_proto_ts_1.InjectiveMegaVaultRpc.GrpcWebError) {
                throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                    code: e.code,
                    context: 'fetchVaultRedemptions',
                    contextModule: this.module,
                });
            }
            throw new exceptions_1.GrpcUnaryRequestException(e, {
                code: exceptions_1.UnspecifiedErrorCode,
                context: 'fetchVaultRedemptions',
                contextModule: this.module,
            });
        }
    }
    async fetchOperatorRedemptionBuckets({ vaultAddress, operatorAddress, }) {
        const request = indexer_proto_ts_1.InjectiveMegaVaultRpc.GetOperatorRedemptionBucketsRequest.create();
        request.vaultAddress = vaultAddress;
        request.operatorAddress = operatorAddress;
        try {
            const response = await this.retry(() => this.client.GetOperatorRedemptionBuckets(request, this.metadata));
            return index_js_2.IndexerGrpcMegaVaultTransformer.operatorRedemptionBucketsResponseToOperatorRedemptionBuckets(response);
        }
        catch (e) {
            if (e instanceof indexer_proto_ts_1.InjectiveMegaVaultRpc.GrpcWebError) {
                throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                    code: e.code,
                    context: 'fetchOperatorRedemptionBuckets',
                    contextModule: this.module,
                });
            }
            throw new exceptions_1.GrpcUnaryRequestException(e, {
                code: exceptions_1.UnspecifiedErrorCode,
                context: 'fetchOperatorRedemptionBuckets',
                contextModule: this.module,
            });
        }
    }
    async fetchVaultTvlHistory({ since, vaultAddress, maxDataPoints, }) {
        const request = indexer_proto_ts_1.InjectiveMegaVaultRpc.TvlHistoryRequest.create();
        request.vaultAddress = vaultAddress;
        request.since = since.toString();
        if (maxDataPoints) {
            request.maxDataPoints = maxDataPoints;
        }
        try {
            const response = await this.retry(() => this.client.TvlHistory(request, this.metadata));
            return index_js_2.IndexerGrpcMegaVaultTransformer.tvlHistoryResponseToTvlHistory(response);
        }
        catch (e) {
            if (e instanceof indexer_proto_ts_1.InjectiveMegaVaultRpc.GrpcWebError) {
                throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                    code: e.code,
                    context: 'fetchVaultTvlHistory',
                    contextModule: this.module,
                });
            }
            throw new exceptions_1.GrpcUnaryRequestException(e, {
                code: exceptions_1.UnspecifiedErrorCode,
                context: 'fetchVaultTvlHistory',
                contextModule: this.module,
            });
        }
    }
    async fetchVaultPnlHistory({ since, vaultAddress, maxDataPoints, }) {
        const request = indexer_proto_ts_1.InjectiveMegaVaultRpc.PnlHistoryRequest.create();
        request.vaultAddress = vaultAddress;
        request.since = since.toString();
        if (maxDataPoints) {
            request.maxDataPoints = maxDataPoints;
        }
        try {
            const response = await this.retry(() => this.client.PnlHistory(request, this.metadata));
            return index_js_2.IndexerGrpcMegaVaultTransformer.pnlHistoryResponseToPnlHistory(response);
        }
        catch (e) {
            if (e instanceof indexer_proto_ts_1.InjectiveMegaVaultRpc.GrpcWebError) {
                throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                    code: e.code,
                    context: 'fetchVaultPnlHistory',
                    contextModule: this.module,
                });
            }
            throw new exceptions_1.GrpcUnaryRequestException(e, {
                code: exceptions_1.UnspecifiedErrorCode,
                context: 'fetchVaultPnlHistory',
                contextModule: this.module,
            });
        }
    }
}
exports.IndexerGrpcMegaVaultApi = IndexerGrpcMegaVaultApi;
