// @injectivelabs/core-proto-ts/esm/injective/evm/v1/access_tuple.js
/* eslint-disable */
import _m0 from "protobufjs/minimal.js";
export const protobufPackage = "injective.evm.v1";
function createBaseAccessTuple() {
    return { address: "", storageKeys: [] };
}
export const AccessTuple = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.address !== "") {
            writer.uint32(10).string(message.address);
        }
        for (const v of message.storageKeys) {
            writer.uint32(18).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAccessTuple();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.address = reader.string();
                    break;
                case 2:
                    message.storageKeys.push(reader.string());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            address: isSet(object.address) ? String(object.address) : "",
            storageKeys: Array.isArray(object?.storageKeys) ? object.storageKeys.map((e) => String(e)) : [],
        };
    },
    toJSON(message) {
        const obj = {};
        message.address !== undefined && (obj.address = message.address);
        if (message.storageKeys) {
            obj.storageKeys = message.storageKeys.map((e) => e);
        }
        else {
            obj.storageKeys = [];
        }
        return obj;
    },
    create(base) {
        return AccessTuple.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseAccessTuple();
        message.address = object.address ?? "";
        message.storageKeys = object.storageKeys?.map((e) => e) || [];
        return message;
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
