// @injectivelabs/core-proto-ts/esm/injective/evm/v1/transaction_logs.js
/* eslint-disable */
import _m0 from "protobufjs/minimal.js";
import { Log } from "./log.js";
export const protobufPackage = "injective.evm.v1";
function createBaseTransactionLogs() {
    return { hash: "", logs: [] };
}
export const TransactionLogs = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.hash !== "") {
            writer.uint32(10).string(message.hash);
        }
        for (const v of message.logs) {
            Log.encode(v, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTransactionLogs();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.hash = reader.string();
                    break;
                case 2:
                    message.logs.push(Log.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            hash: isSet(object.hash) ? String(object.hash) : "",
            logs: Array.isArray(object?.logs) ? object.logs.map((e) => Log.fromJSON(e)) : [],
        };
    },
    toJSON(message) {
        const obj = {};
        message.hash !== undefined && (obj.hash = message.hash);
        if (message.logs) {
            obj.logs = message.logs.map((e) => e ? Log.toJSON(e) : undefined);
        }
        else {
            obj.logs = [];
        }
        return obj;
    },
    create(base) {
        return TransactionLogs.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseTransactionLogs();
        message.hash = object.hash ?? "";
        message.logs = object.logs?.map((e) => Log.fromPartial(e)) || [];
        return message;
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
