// @injectivelabs/networks/dist/cjs/chainInfos.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLocalChainInfo = exports.getDevnetChainInfo = exports.getTestnetChainInfo = exports.getMainnetChainInfo = void 0;
const ts_types_1 = require("@injectivelabs/ts-types");
const INJ_DENOM = 'inj';
const getMainnetChainInfo = () => ({
    feeDenom: INJ_DENOM,
    chainId: ts_types_1.ChainId.Mainnet,
    evmChainId: ts_types_1.EvmChainId.Mainnet,
    env: 'mainnet',
});
exports.getMainnetChainInfo = getMainnetChainInfo;
const getTestnetChainInfo = () => ({
    feeDenom: INJ_DENOM,
    chainId: ts_types_1.ChainId.Testnet,
    evmChainId: ts_types_1.EvmChainId.Sepolia,
    env: 'testnet',
});
exports.getTestnetChainInfo = getTestnetChainInfo;
const getDevnetChainInfo = () => ({
    feeDenom: INJ_DENOM,
    chainId: ts_types_1.ChainId.Devnet,
    evmChainId: ts_types_1.EvmChainId.Sepolia,
    env: 'devnet',
});
exports.getDevnetChainInfo = getDevnetChainInfo;
const getLocalChainInfo = () => ({
    feeDenom: INJ_DENOM,
    chainId: ts_types_1.ChainId.Mainnet,
    evmChainId: ts_types_1.EvmChainId.Mainnet,
    env: 'local',
});
exports.getLocalChainInfo = getLocalChainInfo;
