// @injectivelabs/networks/dist/cjs/network.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isTestnetOrDevnet = exports.isTestnet = exports.isDevnet = exports.isMainnet = exports.getNetworkInfo = exports.getNetworkChainInfo = exports.getChainInfoForNetwork = exports.getNetworkEndpoints = void 0;
const types_js_1 = require("./types.js");
const chainInfos_js_1 = require("./chainInfos.js");
const endpoints_js_1 = require("./endpoints.js");
const getNetworkEndpoints = (network) => {
    switch (network) {
        case types_js_1.Network.MainnetLB:
            return (0, endpoints_js_1.getEndpointsMainnetLB)();
        case types_js_1.Network.MainnetK8s:
            return (0, endpoints_js_1.getEndpointsMainnetK8s)();
        case types_js_1.Network.MainnetSentry:
        case types_js_1.Network.Mainnet:
            return (0, endpoints_js_1.getEndpointsMainnetSentry)();
        case types_js_1.Network.MainnetOld:
            return (0, endpoints_js_1.getEndpointsMainnetOld)();
        case types_js_1.Network.Staging:
            return (0, endpoints_js_1.getEndpointsStaging)();
        case types_js_1.Network.Internal:
            return (0, endpoints_js_1.getEndpointsInternal)();
        case types_js_1.Network.Devnet:
            return (0, endpoints_js_1.getEndpointsDevnet)();
        case types_js_1.Network.Devnet1:
            return (0, endpoints_js_1.getEndpointsDevnet1)();
        case types_js_1.Network.Devnet2:
            return (0, endpoints_js_1.getEndpointsDevnet2)();
        case types_js_1.Network.Devnet3:
            return (0, endpoints_js_1.getEndpointsDevnet3)();
        case types_js_1.Network.TestnetK8s:
            return (0, endpoints_js_1.getEndpointsTestnetK8s)();
        case types_js_1.Network.TestnetOld:
            return (0, endpoints_js_1.getEndpointsTestnetOld)();
        case types_js_1.Network.TestnetSentry:
        case types_js_1.Network.Testnet:
            return (0, endpoints_js_1.getEndpointsTestnetSentry)();
        case types_js_1.Network.Local:
            return (0, endpoints_js_1.getEndpointsLocal)();
        default:
            throw new Error(`Unknown network: ${network}`);
    }
};
exports.getNetworkEndpoints = getNetworkEndpoints;
/**
 * @deprecated - use getNetworkChainInfo instead
 * @param network deprecated
 * @returns
 */
const getChainInfoForNetwork = (network) => (0, exports.getNetworkChainInfo)(network);
exports.getChainInfoForNetwork = getChainInfoForNetwork;
const getNetworkChainInfo = (network) => {
    if (network === types_js_1.Network.Local) {
        return (0, chainInfos_js_1.getLocalChainInfo)();
    }
    if ((0, exports.isTestnet)(network)) {
        return (0, chainInfos_js_1.getTestnetChainInfo)();
    }
    if ((0, exports.isDevnet)(network)) {
        return (0, chainInfos_js_1.getDevnetChainInfo)();
    }
    return (0, chainInfos_js_1.getMainnetChainInfo)();
};
exports.getNetworkChainInfo = getNetworkChainInfo;
const getNetworkInfo = (network) => ({
    ...(0, exports.getNetworkChainInfo)(network),
    ...(0, exports.getNetworkEndpoints)(network),
});
exports.getNetworkInfo = getNetworkInfo;
const isMainnet = (network) => [
    types_js_1.Network.Staging,
    types_js_1.Network.Mainnet,
    types_js_1.Network.MainnetOld,
    types_js_1.Network.MainnetK8s,
    types_js_1.Network.MainnetSentry,
    types_js_1.Network.Internal,
    types_js_1.Network.MainnetLB,
].includes(network);
exports.isMainnet = isMainnet;
const isDevnet = (network) => [
    types_js_1.Network.Devnet,
    types_js_1.Network.Devnet1,
    types_js_1.Network.Devnet2,
    types_js_1.Network.Devnet3,
    types_js_1.Network.Local,
].includes(network);
exports.isDevnet = isDevnet;
const isTestnet = (network) => [
    types_js_1.Network.Testnet,
    types_js_1.Network.TestnetOld,
    types_js_1.Network.TestnetK8s,
    types_js_1.Network.TestnetSentry,
].includes(network);
exports.isTestnet = isTestnet;
const isTestnetOrDevnet = (network) => (0, exports.isDevnet)(network) || (0, exports.isTestnet)(network);
exports.isTestnetOrDevnet = isTestnetOrDevnet;
