// @injectivelabs/networks/dist/cjs/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAssetPriceServiceForNetwork = exports.getPeggyGraphQlEndpointForNetwork = exports.getInjNameReverseResolverContractForNetwork = exports.getInjNameRegistryContractForNetwork = exports.getIncentivesContractForNetwork = exports.getCw20SwapContractForNetwork = exports.getCw20AdapterContractForNetwork = exports.CW20_CODE_IDS_BY_NETWORK = void 0;
const types_js_1 = require("./types.js");
const network_js_1 = require("./network.js");
const CW20_CODE_IDS_BY_NETWORK = (network = types_js_1.Network.Mainnet) => {
    if ((0, network_js_1.isTestnet)(network)) {
        return ['25'];
    }
    return ['28', '5', '42'];
};
exports.CW20_CODE_IDS_BY_NETWORK = CW20_CODE_IDS_BY_NETWORK;
const getCw20AdapterContractForNetwork = (network = types_js_1.Network.Mainnet) => {
    if ((0, network_js_1.isDevnet)(network)) {
        return 'inj1uukt3kqela4vsllvrqnrgllkna5wn3cm588w6k';
    }
    if ((0, network_js_1.isTestnet)(network)) {
        return 'inj1hdvy6tl89llqy3ze8lv6mz5qh66sx9enn0jxg6';
    }
    // mainnet
    return 'inj14ejqjyq8um4p3xfqj74yld5waqljf88f9eneuk';
};
exports.getCw20AdapterContractForNetwork = getCw20AdapterContractForNetwork;
const getCw20SwapContractForNetwork = (network = types_js_1.Network.Mainnet) => {
    if ((0, network_js_1.isDevnet)(network)) {
        return 'inj177yh38g3ctu7cemxpa3c2kvwh2yslfxfmfa66h';
    }
    if ((0, network_js_1.isTestnet)(network)) {
        return 'inj14d7h5j6ddq6pqppl65z24w7xrtmpcrqjxj8d43';
    }
    if (network === types_js_1.Network.Staging) {
        return 'inj12yj3mtjarujkhcp6lg3klxjjfrx2v7v8yswgp9';
    }
    // mainnet
    return 'inj1psk3468yr9teahgz73amwvpfjehnhczvkrhhqx';
};
exports.getCw20SwapContractForNetwork = getCw20SwapContractForNetwork;
const getIncentivesContractForNetwork = (network = types_js_1.Network.Mainnet) => {
    if ((0, network_js_1.isDevnet)(network)) {
        return '';
    }
    if ((0, network_js_1.isTestnet)(network)) {
        return 'inj16twru668nsl7tqzahxd9q033swhr6a5xuslpkt';
    }
    // mainnet
    return '';
};
exports.getIncentivesContractForNetwork = getIncentivesContractForNetwork;
const getInjNameRegistryContractForNetwork = (network = types_js_1.Network.Mainnet) => {
    if ((0, network_js_1.isDevnet)(network)) {
        return 'inj1aw59rkpd9afp2ws6rx23nz5mrvq8dlckeslwfa';
    }
    if ((0, network_js_1.isTestnet)(network)) {
        return 'inj1aw59rkpd9afp2ws6rx23nz5mrvq8dlckeslwfa';
    }
    // mainnet
    return 'inj1hm8vs8sr2h9nk0x66vctfs528wrp6k3gtgg275';
};
exports.getInjNameRegistryContractForNetwork = getInjNameRegistryContractForNetwork;
const getInjNameReverseResolverContractForNetwork = (network = types_js_1.Network.Mainnet) => {
    if ((0, network_js_1.isDevnet)(network)) {
        return 'inj1knf6puyscuuqqhgqglskfc0k99d4885qw5uv7v';
    }
    if ((0, network_js_1.isTestnet)(network)) {
        return 'inj1knf6puyscuuqqhgqglskfc0k99d4885qw5uv7v';
    }
    return 'inj1x9m0hceug9qylcyrrtwqtytslv2jrph433thgu';
};
exports.getInjNameReverseResolverContractForNetwork = getInjNameReverseResolverContractForNetwork;
const getPeggyGraphQlEndpointForNetwork = (network) => {
    if ((0, network_js_1.isDevnet)(network)) {
        return 'https://api.thegraph.com/subgraphs/name/injectivelabsdev/injective-peggo-devnet';
    }
    if ((0, network_js_1.isTestnet)(network)) {
        return 'https://api.thegraph.com/subgraphs/name/injectivelabs/injective-peggo-goerli';
    }
    // mainnet
    return 'https://api.thegraph.com/subgraphs/name/injectivelabs/injective-peggo-mainnet';
};
exports.getPeggyGraphQlEndpointForNetwork = getPeggyGraphQlEndpointForNetwork;
const getAssetPriceServiceForNetwork = (network) => {
    if ((0, network_js_1.isDevnet)(network)) {
        return 'https://devnet.asset.injective.dev/asset-price/v1';
    }
    if ((0, network_js_1.isTestnet)(network)) {
        return 'https://k8s.testnet.asset.injective.network/asset-price/v1';
    }
    // mainnet
    return 'https://k8s.mainnet.asset.injective.network/asset-price/v1';
};
exports.getAssetPriceServiceForNetwork = getAssetPriceServiceForNetwork;
