// @injectivelabs/core-proto-ts/cjs/injective/peggy/v1/ethereum_signer.js
"use strict";
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignType = exports.protobufPackage = void 0;
exports.signTypeFromJSON = signTypeFromJSON;
exports.signTypeToJSON = signTypeToJSON;
exports.protobufPackage = "injective.peggy.v1";
/** SignType defines messages that have been signed by an orchestrator */
exports.SignType = {
    SIGN_TYPE_UNKNOWN: 0,
    SIGN_TYPE_ORCHESTRATOR_SIGNED_MULTI_SIG_UPDATE: 1,
    SIGN_TYPE_ORCHESTRATOR_SIGNED_WITHDRAW_BATCH: 2,
    UNRECOGNIZED: -1,
};
function signTypeFromJSON(object) {
    switch (object) {
        case 0:
        case "SIGN_TYPE_UNKNOWN":
            return exports.SignType.SIGN_TYPE_UNKNOWN;
        case 1:
        case "SIGN_TYPE_ORCHESTRATOR_SIGNED_MULTI_SIG_UPDATE":
            return exports.SignType.SIGN_TYPE_ORCHESTRATOR_SIGNED_MULTI_SIG_UPDATE;
        case 2:
        case "SIGN_TYPE_ORCHESTRATOR_SIGNED_WITHDRAW_BATCH":
            return exports.SignType.SIGN_TYPE_ORCHESTRATOR_SIGNED_WITHDRAW_BATCH;
        case -1:
        case "UNRECOGNIZED":
        default:
            return exports.SignType.UNRECOGNIZED;
    }
}
function signTypeToJSON(object) {
    switch (object) {
        case exports.SignType.SIGN_TYPE_UNKNOWN:
            return "SIGN_TYPE_UNKNOWN";
        case exports.SignType.SIGN_TYPE_ORCHESTRATOR_SIGNED_MULTI_SIG_UPDATE:
            return "SIGN_TYPE_ORCHESTRATOR_SIGNED_MULTI_SIG_UPDATE";
        case exports.SignType.SIGN_TYPE_ORCHESTRATOR_SIGNED_WITHDRAW_BATCH:
            return "SIGN_TYPE_ORCHESTRATOR_SIGNED_WITHDRAW_BATCH";
        case exports.SignType.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
