// @audius/eth/dist/cjs/governance/governance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Governance = void 0;
const abi_1 = require("./abi");
const constants_1 = require("./constants");
class Governance {
    constructor(client, { address } = {}) {
        this.getVotingPeriod = () => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getVotingPeriod'
        });
        this.getVotingQuorumPercent = () => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getVotingQuorumPercent'
        });
        this.getExecutionDelay = () => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getExecutionDelay'
        });
        this.getProposalById = ({ id }) => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getProposalById',
            args: [id]
        });
        this.getProposalTargetContractHash = ({ id }) => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getProposalTargetContractHash',
            args: [id]
        });
        this.getInProgressProposals = () => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getInProgressProposals'
        });
        this.getVoteInfoByProposalAndVoter = ({ id, voterAddress }) => this.client.readContract({
            address: this.address,
            abi: abi_1.abi,
            functionName: 'getVoteInfoByProposalAndVoter',
            args: [id, voterAddress]
        });
        this.getProposalSubmittedEvents = ({ fromBlock = BigInt(0), proposalId, proposer }) => {
            const args = {};
            if (proposalId)
                args._proposalId = proposalId;
            if (proposer)
                args._proposer = proposer;
            return this.client.getContractEvents({
                address: this.address,
                abi: abi_1.abi,
                eventName: 'ProposalSubmitted',
                fromBlock,
                args
            });
        };
        this.getProposalOutcomeEvaluatedEvents = ({ fromBlock = BigInt(0), proposalId }) => {
            const args = {};
            if (proposalId)
                args._proposalId = proposalId;
            return this.client.getContractEvents({
                address: this.address,
                abi: abi_1.abi,
                eventName: 'ProposalOutcomeEvaluated',
                fromBlock,
                args
            });
        };
        this.getProposalVoteSubmittedEvents = ({ fromBlock = BigInt(0), proposalId, voter }) => {
            const args = {};
            if (proposalId)
                args._proposalId = proposalId;
            if (voter)
                args._voter = voter;
            return this.client.getContractEvents({
                address: this.address,
                abi: abi_1.abi,
                eventName: 'ProposalVoteSubmitted',
                fromBlock,
                args
            });
        };
        this.getProposalVoteUpdatedEvents = ({ fromBlock = BigInt(0), proposalId, voter }) => {
            const args = {};
            if (proposalId)
                args._proposalId = proposalId;
            if (voter)
                args._voter = voter;
            return this.client.getContractEvents({
                address: this.address,
                abi: abi_1.abi,
                eventName: 'ProposalVoteUpdated',
                fromBlock,
                args
            });
        };
        this.client = client;
        this.address = address !== null && address !== void 0 ? address : constants_1.GOVERNANCE_CONTRACT_ADDRESS;
    }
}
exports.Governance = Governance;
